---
title: "CHIKV LNSC figures"
author: "Ryan Sheridan"
date: "`r Sys.Date()`"
output: 
  html_document:
    toc:       true
    toc_float: true
    toc_depth: 2
    theme:     cosmo
    highlight: kate
params:
  template_dir: "src"
  so_dir:       "~/Dropbox/Ryan/Projects/morrison-lnsc/results/sobjs"
  orig_so_dir:  "~/Dropbox/Ryan/Projects/morrison-scRNA-seq/results"
editor_options: 
  chunk_output_type: console
---

<br>


```{r "chunk opts", echo = FALSE}
# Default chunk options
knitr::opts_chunk$set(
  message = FALSE,
  warning = FALSE,
  echo    = FALSE,
  dpi     = 500
)
```

```{r "packages"}
# tidyverse
library(tidyverse)
library(here)
library(knitr)
library(broom)

# plotting
library(scales)
library(ggrepel)
library(ggtrace)
library(patchwork)
library(cowplot)
library(colorblindr)
library(ggtext)

# analysis
library(Seurat)
library(clustifyr)
library(clustifyrdata)
library(DoubletFinder)
library(M3Drop)
library(CellChat)
library(djvdj)
library(presto)
library(gprofiler2)
library(qs)

library(clusterProfiler)
library(enrichplot)
library(msigdbr)

source(here(params$template_dir, "funs.R"))

chikv_grps <- c("CHIKV-low", "CHIKV-high")
```

```{r "functions"}
type_lab_fn <- function(x) str_replace(x, "_", " ")

plot_example_gene <- function(so_in, feat, grp, box_colors = NULL,
                              umap_colors = c("#56B4E9", "white", "#D7301F")) {
  
  dat <- so_in %>%
    FetchData(c(feat, grp, "UMAP_1", "UMAP_2"))
  
  u <- dat %>%
    arrange(!!sym(feat)) %>%
    ggplot(aes(UMAP_1, UMAP_2, fill = !!sym(feat))) +
    geom_point_trace(size = 0.25, stroke = 0.65) +
    scale_fill_gradientn(colours = umap_colors) +
    guides(fill = guide_colorbar(
      ticks          = FALSE,
      barheight      = unit(10, "pt"),
      barwidth       = unit(140, "pt"),
      title.position = "top"
    )) +
    umap_theme +
    theme(
      legend.position      = "top",
      legend.title         = element_text(hjust = 0.5),
      legend.justification = 0.5,
      panel.border         = element_rect(fill = NA, color = "grey85")
    )
  
  
  lvls <- dat %>%
    group_by(!!sym(grp)) %>%
    summarize(
      med   = median(!!sym(feat)),
      stats = list(boxplot.stats(!!sym(feat))),
      q3    = map_dbl(stats, ~ .x$stats[4]),
      q4    = map_dbl(stats, ~ .x$stats[5]),
      max   = max(!!sym(feat)),
      .groups = "drop"
    ) %>%
    arrange(med, q3, q4, max) %>%
    pull(grp) %>%
    rev()
  
  b <- dat %>%
    plot_violin(
      feat, grp,
      plot_colors   = box_colors,
      plot_lvls     = lvls,
      method        = "boxplot",
      outlier.size  = 0.25,
      outlier.alpha = 1
    ) +
    theme(
      legend.position = "none",
      aspect.ratio = 0.21,
      axis.text.x  = element_text(angle = 45, hjust = 1),
      axis.line.y  = element_line(size = 0.5, color = "grey85"),
      axis.ticks.x = element_blank(),
      panel.border = element_blank(),
      axis.title.x = element_blank()
    )
  
  plot_grid(
    u, b,
    ncol        = 1,
    rel_heights = c(1, 0.5),
    align       = "v",
    axis        = "rl"
  )
}

create_cor_heatmap <- function(df_in, ttl, lec_labs = NULL, ln_clr = fade_0) {
  
  # # Cell type labels
  # lec_labs <- df_in %>%
  #   format_labels(
  #     lab_clmn = "lec_type",
  #     char_vec = new_cell_labs,
  #     expr_vec = marco_expr
  #   )
  # 
  # ref_labs <- c(
  #   Valve      = "Valve LEC",
  #   Collecting = "Collecting LEC",
  #   Ptx3_LEC   = "Ptx3 LEC",
  #   Marco_LEC  = parse(text = marco_expr)
  # )
  
  if (!is.null(lec_labs)) {
    df_in <- df_in %>%
      filter(assigned_type %in% names(lec_labs()))
  }
  
  # Data for tile labels
  lab_df <- df_in %>%
    group_by(assigned_type) %>%
    slice_max(r, n = 1) %>%
    ungroup()
  
  # Create heatmaps
  res <- df_in %>%
    ggplot(aes(ref_type, assigned_type, fill = r)) +
    geom_tile(color = ln_clr, size = 0.2) +
    
    geom_text(
      aes(label = round(r, 2)),
      data  = lab_df,
      color = "black",
      size  = txt_pt1 / .pt
    ) +
    
    guides(fill = bar_gd()) +
    
    scale_fill_gradientn(colors = c(fade_0, "#E69F00")) +
    # scale_x_discrete(labels = ref_labs) +
    # scale_y_discrete(labels = lec_labs) +
    
    labs(title = ttl, x = "reference", y = "assigned type") +
    
    base_theme +
    theme(
      legend.text = element_text(size = txt_pt1),
      axis.line.x = element_blank(),
      axis.line.y = element_blank(),
      axis.text.x = element_text(hjust = 1, angle = 45, size = txt_pt1),
      axis.text.y = element_text(hjust = 1, size = txt_pt1),
      axis.title  = element_text(size = txt_pt2),
      plot.title  = element_text(size = ttl_pt1)
    )
  
  if (!is.null(lec_labs)) {
    res <- res +
      scale_y_discrete(labels = lec_labs)
  }
  
  res
}

create_gene_boxplot <- function(input, GENE, TYPE, clr = lec_type_cols[[TYPE]],
                                pval_y = 5) {
  
  # Format plot data
  dat <- input %>%
    FetchData(c("lec_type", "treatment", "rep", "orig.ident", "chikv_grp", GENE)) %>%
    filter(lec_type == TYPE) %>%
    mutate(chikv_grp = ifelse(treatment == "mock", treatment, chikv_grp))
  
  clrs <- set_names(
    c("grey", rep(clr, 2)),
    unique(dat$chikv_grp)
  )
  
  # Calculate p-values
  p_vals <- dat %>%
    calc_p_vals(
      data_column = GENE,
      type_column = "chikv_grp"
    )
  
  comps <- names(clrs) %>%
    combn(2, simplify = FALSE)
  
  comps <- list(
    x    = map_chr(comps, pluck, 1),
    xend = map_chr(comps, pluck, 2),
    y    = c(pval_y, pval_y + pval_y * 0.12, pval_y)
  )
  
  # Create boxplots
  res <- dat %>%
    plot_violin(
      GENE, "chikv_grp",
      method        = "boxplot",
      plot_colors   = clrs,
      plot_lvls     = names(clrs),
      alpha         = c(0.5, 0.1, 0.5),
      size          = 1,
      outlier.size  = 1.5,
      outlier.alpha = 1,
      width         = 0.6
    ) +
    labs(title = TYPE, y = str_c(GENE, " expression")) +
    base_theme +
    theme(
      legend.position = "none",
      aspect.ratio = 1.5,
      plot.title   = element_text(size = ttl_pt2 * 1.5),
      axis.title.y = element_text(size = ttl_pt2),
      axis.title.x = element_blank(),
      axis.text.x  = element_text(size = ttl_pt2, hjust = c(0.75, 0.5, 0.25)),
      plot.margin  = margin(5, 10, 5, 5)
    )
  
  # Add p-values to plot
  comps %>%
    pwalk(~ {
      v_args <- list(...)[c("x", "xend")]
      
      p <- p_vals %>%
        filter(`Cell type 1` %in% v_args & `Cell type 2` %in% v_args) %>%
        pull(p_adj)
      
      res <<- res %>%
        add_pvals(
          ...,
          p_val    = p,
          size     = txt_pt2 / .pt,
          line_col = "grey75"
        )
    })
  
  res
}

create_example_gene_fig <- function(df_in, gns, x_clmn, clrs = lec_type_cols,
                                    ttl = NULL) {
  res <- gns %>%
    imap(~ {
      lvls <- df_in %>%
        group_by(!!sym(x_clmn), tm_treat) %>%
        summarize(
          n     = n(),
          stats = list(boxplot.stats(!!sym(.x))),
          med   = map_dbl(stats, ~ .x$stats[3]),
          q3    = map_dbl(stats, ~ .x$stats[4]),
          q4    = map_dbl(stats, ~ .x$stats[5]),
          mx    = max(!!sym(.x)),
          .groups = "drop"
        ) %>%
        group_by(!!sym(x_clmn)) %>%
        summarize(across(c(med, q3, q4, mx), mean)) %>%
        arrange(med, q3, q4, mx) %>%
        pull(x_clmn) %>%
        rev()
      
      plt <- df_in %>%
        mutate(
          x_type    = fct_relevel(!!sym(x_clmn), lvls),
          chikv_grp = fct_relevel(chikv_grp, chikv_grps)
        ) %>%
        ggplot(
          # aes(x_type, !!sym(.x), color = x_type, fill = x_type, alpha = chikv_grp)
          aes(x_type, !!sym(.x), color = x_type, fill = x_type)
        ) +
        geom_boxplot(
          linewidth = 0.5, alpha = 0.5,  # color = "black",
          outlier.size = 0.01, outlier.alpha = 0.1,
          position = position_dodge2(preserve = "single"),
          key_glyph = draw_key_point,
        ) +
        scale_color_manual(values = clrs) +
        scale_fill_manual(values = clrs) +
        scale_alpha_manual(values = c(`CHIKV-low` = 0.5, `CHIKV-high` = 1)) +
        scale_x_discrete(labels = type_lab_fn) +
        facet_wrap(~ tm_treat, ncol = 1) +
        labs(y = str_c(.x, " expression")) +
        guides(
          fill = "none", color = "none",
          alpha = guide_legend(
            override.aes = list(shape = 15, fill = "black", size = 4)
          )
        ) +
        # base_theme +
        djvdj_theme() +
        theme(
          aspect.ratio    = 0.15,
          plot.title      = element_text(hjust = 0.5),
          legend.position = "top",
          legend.title    = element_blank(),
          # legend.text     = element_text(size = txt_pt1),
          axis.title.x    = element_blank(),
          axis.text.x     = element_text(angle = 45, hjust = 1),
          # axis.text.y     = element_text(size = txt_pt1 * 0.75),
          # strip.text      = element_text(size = txt_pt2)
        )
      
      if (.y == 1 && !is.null(ttl)) {
        plt <- plt + ggtitle(ttl)
      }
      
      plt
    })
  
  res
}

.format_pvalue <- function(p, digits = 1, cutoffs = NULL) {

  # Set p label based on vector of cutoffs
  if (!is.finite(p)) return(as.character(NA))

  if (!is.null(cutoffs)) {
    if (any(duplicated(cutoffs))) {
      cli::cli_abort("Cutoff values for p_label must be unique.")
    }

    # Set default labels when not provided by user
    if (is.null(names(cutoffs))) {
      cutoffs <- sort(cutoffs, decreasing = TRUE)

      names(cutoffs) <- purrr::imap_chr(
        cutoffs, ~ paste0(rep("*", .y), collapse = "")
      )
    }

    cutoffs <- sort(cutoffs)
    p_label <- as.character(NA)

    for (val in names(cutoffs)) {
      if (p < cutoffs[val]) {
        p_label <- val

        break()
      }
    }

    # Treat "value" as a keyword that will allow user to display actual
    # p-value for a certain cutoff
    # All custom labels need to be wrapped in quotes for parsing
    if (!identical(p_label, "value")) {
      if (!is.na(p_label)) p_label <- paste0("\'", p_label, "\'")

      return(p_label)
    }
  }

  # Format p-value label
  if (p >= 0.1)    return(as.character(round(p, 1)))
  else if (p == 0) return(as.character(p))

  p <- scales::label_scientific(digits = digits)(p)

  ex <- str_extract_all(p, "[+\\-][0-9]+$")

  p <- sub(paste0("\\", ex, "$"), "", p)

  ex <- as.numeric(ex)
  ex <- as.character(ex)

  p <- sub("e", "*x*10^", p)
  p <- paste0(p, ex)

  p
}

create_example_gene_fig2 <- function(df_in, gns, x_clmn, al_clmn,
                                     plot_lvls = NULL, clrs = lec_type_cols,
                                     ttl = NULL) {
  
  vars <- unique(df_in[[al_clmn]])
  
  p_cuts <- c(0.01, 0.001, 0.0001)
  
  res <- gns %>%
    imap(~ {
      # Calculate p-values
      # ONLY SHOW <0.01
      p_dat <- df_in %>%
        mutate(x_type = !!sym(x_clmn)) %>%
        group_by(x_type) %>%
        filter(n_distinct(!!sym(al_clmn)) == 2) %>%
        summarize(
          p = wilcox.test(
            (!!sym(.x))[!!sym(al_clmn) == vars[1]],
            (!!sym(.x))[!!sym(al_clmn) == vars[2]]
          )$p.value,
          .groups = "drop"
        ) %>%
        filter(!is.na(p)) %>%
        mutate(p_adj = p.adjust(p)) %>%
        rowwise() %>%
        mutate(p_lab = .format_pvalue(p_adj, cutoffs = p_cuts)) %>%
        ungroup()
        # filter(p_adj < 0.01)
      
      # Set plot levels
      lvls <- df_in %>%
        group_by(!!sym(x_clmn), tm_treat) %>%
        summarize(
          n     = n(),
          stats = list(boxplot.stats(!!sym(.x))),
          med   = map_dbl(stats, ~ .x$stats[3]),
          q3    = map_dbl(stats, ~ .x$stats[4]),
          q4    = map_dbl(stats, ~ .x$stats[5]),
          mx    = max(!!sym(.x)),
          .groups = "drop"
        ) %>%
        group_by(!!sym(x_clmn)) %>%
        summarize(across(c(med, q3, q4, mx), mean)) %>%
        arrange(med, q3, q4, mx) %>%
        pull(x_clmn) %>%
        rev()
      
      lvls <- unique(c(plot_lvls, lvls))
      
      # Create boxplots
      plt <- df_in %>%
        mutate(
          x_type    = fct_relevel(!!sym(x_clmn), lvls),
          chikv_grp = fct_relevel(chikv_grp, chikv_grps),
          treatment = fct_relevel(treatment, treats)
        ) %>%
        ggplot(
          aes(x_type, !!sym(.x), color = x_type, fill = x_type, alpha = !!sym(al_clmn))
        ) +
        geom_boxplot(
          linewidth = 0.5,
          outlier.size = 0.2, outlier.alpha = 0.1,
          position = position_dodge2(preserve = "single"),
          key_glyph = draw_key_point,
        ) +
        scale_color_manual(values = clrs) +
        scale_fill_manual(values = clrs) +
        scale_alpha_manual(values = c(mock = 0.25, CHIKV = 0.75)) +
        scale_x_discrete(labels = type_lab_fn) +
        facet_wrap(~ tm, ncol = 1) +
        labs(y = str_c(.x, " expression")) +
        guides(
          fill = "none", color = "none",
          alpha = guide_legend(
            override.aes = list(shape = 15, fill = "black", size = 4)
          )
        ) +
        djvdj_theme() +
        theme(
          # aspect.ratio    = 0.25,
          plot.title      = element_text(hjust = 0.5),
          legend.position = "top",
          legend.title    = element_blank(),
          # legend.text     = element_text(size = txt_pt1),
          axis.title.x    = element_blank(),
          axis.text.x     = element_text(angle = 45, hjust = 1),
          # axis.text.y     = element_text(size = txt_pt1 * 0.75),
          strip.text      = element_text(size = txt_pt2)
        )
      
      # Add p-values
      plt <- plt +
        geom_text(
          aes(y = Inf, fill = NULL, label = p_lab),
          data  = p_dat,
          vjust = 1.5,
          color = "black",
          alpha = 1,
          parse = TRUE,
          size  = 16 / .pt
        ) +
        scale_y_continuous(expand = expansion(c(0.05, 0.15)))
      
      if (.y == 1 && !is.null(ttl)) {
        plt <- plt + ggtitle(ttl)
      }
      
      plt
    })
  
  res
}
```

```{r "load objects"}

so_8_dir  <- here(params$so_dir, "so_8.qs")
so_24_dir <- here(params$so_dir, "so_24.qs")

# Replace this with setup.Rmd that has code used to process each sample
if (!file.exists(so_8_dir) || !file.exists(so_24_dir)) {
  so_8  <- read_rds(file.path(params$orig_so_dir, "2021-07-16/sobjs/so.rds"))
  so_24 <- read_rds(file.path(params$orig_so_dir, "2021-01-08/sobjs/so.rds"))
  
  # Adjust 24 hpi
  # Add PvCs to cell types
  # This is based on clustifyr results using ref_lymphnodestromal, and Pecam1,
  # Pdpn, Itga7, and Acta2 expression
  lec_cell_types <- "Endothelial cells"
  fib_cell_types <- c("Fibroblasts", "Stromal cells")
  
  so_24 <- so_24 %>%
    mutate_meta(
      mutate,
      cell_type   = case_when(
        fib_type == "PvC"             ~ fib_type,
        cell_type %in% fib_cell_types ~ "FRC",
        TRUE                          ~ cell_type
      )
    )
  
  so_24 <- so_24 %>%
    mutate_meta(
      mutate,
      lec_type = case_when(
        lec_subtype == "unassigned"   ~ "unassigned-LEC",
        cell_type %in% lec_cell_types ~ lec_subtype,
        TRUE                          ~ cell_type
      )
    )
  
  so_24 <- so_24 %>%
    mutate_meta(
      mutate,
      chikv_type = if_else(chikv_grp == chikv_grps[2], lec_type, chikv_grp)
    )
  
  # Adjust 8 hpi
  # Cluster 27 appears to be FRCs based on Pdpn, Pecam1, and Pdgfra
  so_8 <- so_8 %>%
    mutate_meta(
      mutate, 
      orig.ident = str_c(treatment, "-", rep),
      lec_type = recode(
        lec_type,
        `unassigned-LEC`     = "unassigned",
        Fibroblasts          = "FRC",
        `Stromal cells (DN)` = "PvC"
      ),
      chikv_grp = ifelse(tot_nCount_CHIKV > 0, chikv_grps[2], chikv_grps[1])
    )
  
  so_8 <- so_8 %>%
    mutate_meta(
      mutate,
      lec_type = ifelse(RNA_snn_res.5 == "27", "FRC", lec_type)
    )
  
  # Save objects
  qsave(so_8, so_8_dir)
  qsave(so_24, so_24_dir)
  
  rm(so_8, so_24)
}

# # Merged objects
# so_merge_dir <- file.path(params$so_dir, "so_merge.qs")
# 
# if (!file.exists(so_merge_dir)) {
#   so_8  <- qread(so_8_dir)
#   so_24 <- qread(so_24_dir)
#   
#   so_merge <- merge(so_8, so_24)
#   
#   rm(so_8, so_24)
#   gc()
#   
#   so_merge <- so_merge %>%
#     ScaleData(block.size = 200)
#   
#   so_merge <- so_merge %>%
#     FindVariableFeatures(nFeatures = 2000) %>%
#     RunPCA() %>%
#     RunUMAP(dims = 1:40)
#   
#   qsave(so_merge, file.path(params$so_dir, "so_merge.qs"))
#  
# } else {
#   so_merge <- qread(so_merge)
# }

so_8  <- qread(so_8_dir)
so_24 <- qread(so_24_dir)

so_merge_df <- bind_rows(so_8@meta.data, so_24@meta.data)

# Load LEC objects
so_lec_8  <- read_rds(file.path(params$orig_so_dir, "2021-07-16/sobjs/so_lec.rds"))
so_lec_24 <- read_rds(file.path(params$orig_so_dir, "2021-01-08/sobjs/so_lec.rds"))

# Set levels
sam_lvls   <- unique(so_24$orig.ident)
treats     <- unique(so_24$treatment)
```

```{r "references"}
# LEC reference
ref_lec <- ref_immgen[, grepl("^Endothelial cells", colnames(ref_immgen))]
ref_lec <- ref_lec[rownames(ref_lec) %in% rownames(ref_LEC_xiang), ]

colnames(ref_lec) <- colnames(ref_lec) %>%
  str_replace("Endothelial cells \\(BEC\\)", "BEC")

ref_LEC_xiang <- ref_LEC_xiang[rownames(ref_lec), ]

if (!identical(rownames(ref_LEC_xiang), rownames(ref_lec))) {
  stop("LEC reference rownames do not match.")
}

ref_lec <- cbind(ref_LEC_xiang, ref_lec)
ref_lec <- ref_lec[, !grepl("^Endothelial cells", colnames(ref_lec))]

# Fibroblast/stromal cell reference
ref_fib <- ref_immgen[, grepl("^Fibroblast", colnames(ref_immgen))]
ref_fib <- ref_fib[rownames(ref_fib) %in% rownames(ref_lymphnodestromal), ]

ref_lymphnodestromal <- ref_lymphnodestromal[rownames(ref_fib), ]

if (!identical(rownames(ref_lymphnodestromal), rownames(ref_fib))) {
  stop("Fibroblast/stromal reference rownames do not match.")
}

ref_fib <- cbind(ref_lymphnodestromal, ref_fib)
```

```{r "theme"}
# ggplot2 themes
txt_pt1  <- 10
txt_pt2  <- 12
ttl_pt1  <- 14
ttl_pt2  <- 16
ln_pt    <- 0.5
ln_col   <- "grey85"
cir_size <- 3.5
sqr_size <- 4

txt_theme_1 <- theme(
  strip.text  = element_text(size = ttl_pt2),
  legend.text = element_text(size = txt_pt2),
  axis.title  = element_text(size = ttl_pt2),
  axis.text   = element_text(size = txt_pt2)
)

line_theme <- theme(
  axis.line.x  = element_line(linewidth = ln_pt, color = ln_col),
  axis.line.y  = element_line(linewidth = ln_pt, color = ln_col),
  axis.ticks.x = element_line(linewidth = ln_pt, color = ln_col),
  axis.ticks.y = element_line(linewidth = ln_pt, color = ln_col)
)

base_theme <- djvdj_theme() +
  theme(
    plot.title       = element_text(face = "plain", size = ttl_pt2),
    strip.background = element_blank(),
    strip.text       = element_text(face = "plain")
  ) +
  txt_theme_1

umap_theme <- base_theme +
  theme(
    panel.border = element_rect(color = NA, fill = NA),
    axis.title   = element_blank(),
    axis.line.x  = element_blank(),
    axis.line.y  = element_blank(),
    axis.ticks.x = element_blank(),
    axis.ticks.y = element_blank(),
    axis.text    = element_blank()
  )

fade_0 <- "#FAFAFA"
fade_1 <- "#F0F0F0"
fade_2 <- "#D9D9D9"

hist_y_lab <- "number of cells"

# alpha for plots
al <- 0.7

# Okabe Ito color palettes
ito_cols <- c(
  palette_OkabeIto[c(1:3, 5)], "#D7301F",
  palette_OkabeIto[c(4, 6)],   "#6A51A3",
  palette_OkabeIto[7],         "#875C04",
  "#065D43", "#F7AF34", "#00446E"
)

ito_cols <- ito_cols[3:length(ito_cols)] %>%
  darken(0.2) %>%
  c(ito_cols, ., "#686868", "#000000")

# Set sample colors
get_cols <- create_col_fun(ito_cols)

sam_cols <- c(
  # "#F03B20", "#FD8D3C",
  "#BD0026", "#FECC5C", "#FFFFB2",
  # "#0072B2", "#56B4E9",
  "#00446E", "#035B8F", "#2A8FBF"
)

names(sam_cols) <- sam_lvls

# CHIKV treatment groups
chikv_infctd <- so_merge_df %>%
  group_by(treatment) %>%
  summarize(mn = mean(nCount_CHIKV), .groups = "drop") %>%
  filter(mn == max(mn)) %>%
  pull(treatment)

# Cell type colors
type_cols <- unique(so_merge_df$cell_type)
type_cols <- set_names(ito_cols[seq_along(type_cols)], type_cols)

type_cols["unassigned"] <- "#999999"

subtype_cols <- c(colnames(ref_lec), colnames(ref_fib))
subtype_cols <- set_names(ito_cols[seq_along(subtype_cols)], subtype_cols)
subtype_cols["unassigned"] <- "#999999"
subtype_cols["other"]      <- fade_0
subtype_cols["CD34+ SC"]   <- "#534379"

lec_type_cols <- c(subtype_cols, type_cols[names(type_cols) != "unassigned"])
lec_type_cols["CHIKV-low"]      <- "white"
lec_type_cols["T cells"]        <- "black"
lec_type_cols["B cells"]        <- "#875C04"
lec_type_cols["FRC"]            <- "#00375B"
lec_type_cols["PvC"]            <- "#065D43"
lec_type_cols["unassigned-LEC"] <- "#676767"


lec_type_cols <- so_merge_df %>%
  pull(lec_type) %>%
  table() %>%
  sort(decreasing = TRUE) %>%
  names()

lec_type_cols <- lec_type_cols[lec_type_cols != "unassigned"]

lec_type_cols <- set_names(
  ito_cols[seq_along(lec_type_cols)],
  lec_type_cols
)

lec_type_cols["unassigned"] <- "grey70"

old_cols <- lec_type_cols

lec_type_cols["FRC"]   <- old_cols["Valve"]
lec_type_cols["Valve"] <- old_cols["fLEC"]
lec_type_cols["fLEC"]  <- old_cols["FRC"]

# CHIKV clusters colors
grp_cols <- set_names(
  c("#56B4E9", "#0072B2"),
  chikv_grps
)

n_reps <- n_distinct(so_merge_df$rep)

grp_rep_cols <- grp_cols %>% 
  imap(~ set_names(
    rep(.x, n_reps),
    str_c(.y, "-", 1:n_reps)
  )) %>%
  flatten_chr()

# Vectors to modify cell type labels for plotting
new_cell_labs <- c(
  Valve      = "Valve LEC",
  Collecting = "Collecting LEC",
  Ptx3_LEC   = "Ptx3 LEC"
)

marco_expr <- c(Marco_LEC = "\"Marco\"^\"+\"*\" LEC\"")
```

```{r "8 dpi DEGs"}

# Identify 8 dpi DEGs
Idents(so_8) <- so_8$treatment

typs <- get_cell_types(so_8, "subtype", "orig.ident")
typs <- typs[typs != "other"]

degs <- typs %>%
  map_dfr(~ {
    so_8 %>%
      subset(subtype == .x) %>%
      FindConservedMarkers(
        ident.1         = "CHIKV",
        ident.2         = "mock",
        grouping.var    = "rep",
        logfc.threshold = 0.1,
      ) %>%
      mutate(subtype = .x) %>%
      as_tibble(rownames = "gene")
  })

# Add summary stats
reps <- unique(so_8$rep)

fc_clmns   <- str_c(reps, "_avg_log2FC")
pct1_clmns <- str_c(reps, "_pct.1")
pct2_clmns <- str_c(reps, "_pct.2")

top_degs <- degs %>%
  filter(
    if_all(all_of(fc_clmns), ~ .x > 0.15),
    max_pval       < 0.05
  ) %>%
  split(.$subtype) %>%
  map(pull, gene)

# go <- top_degs %>%
#   gost(organism = "mmusculus")
# 
# go_df <- go$result %>%
#   filter(
#     term_size > 10,
#     term_size < 1000,
#     p_value   < 0.05
#   )

```

```{r "8dpi GO"}

# GO tree
library(clusterProfiler)
library(org.Mm.eg.db)
library(enrichplot)
library(GOSemSim)
library(DOSE)

go_new <- top_degs %>%
  compareCluster(
    fun          = enrichGO,
    keyType      = "SYMBOL",
    OrgDb        = org.Mm.eg.db,
    maxGSSize    = 1000,
    minGSSize    = 10,
    pvalueCutoff = 1.1,
    qvalueCutoff = 1.1,
    ont          = "BP"
  ) %>%
  simplify()

# Most frequent GO terms
top_terms <- go_new@compareClusterResult %>%
  dplyr::filter(p.adjust < 0.05, qvalue < 0.2) %>%
  group_by(Description) %>%
  summarize(
    n        = n_distinct(Cluster),
    p.adjust = mean(p.adjust),
    qvalue   = mean(qvalue)
  ) %>%
  arrange(desc(n))

```

```{r "GO clusters"}

#' Fill the upper triangular matrix completely
#'
#' @param x termsim similarity matrix
#' @param keep GO terms to keep
#' @return a data.frame
#' @noRd
fill_termsim <- function(x, keep = NULL) {
  if (!is.null(keep)) termsim <- x[keep, keep]
  else                termsim <- x
  
  termsim[which(is.na(termsim))] <- 0
  
  termsim2 <- termsim + t(termsim)
  
  for ( i in seq_len(nrow(termsim2))) termsim2[i, i] <- 1
  
  termsim2
}

go_dist <- go_new %>%
  filter(p.adjust < 0.05, qvalue < 0.2) %>%
  pairwise_termsim() %>%
  .@termsim

clst_terms <- top_terms %>%
  filter(n > 1) %>%
  pull(Description)

go_dist <- go_dist %>%
  fill_termsim(keep = clst_terms)

# Use the ward.D method to avoid overlapping ancestor nodes of each group
hc <- stats::hclust(
  stats::as.dist(1 - go_dist),
  method = "ward.D"
)

clsts <- stats::cutree(hc, 3)

clsts <- tibble(
  Description = names(clsts),
  cluster     = unname(clsts)
)

# Create heatmap
# top_10 <- top_terms %>%
#   head(10) %>%
#   pull(Description)

# heat_dat <- go_new@compareClusterResult %>%
#   left_join(clsts, by = "Description")
  # filter(Description %in% top_10) %>%
  # mutate(Description = fct_relevel(Description, rev(top_10))) %>%
  # arrange(Description) %>%
  # mutate(
  #   Description = str_trunc(as.character(Description), width = 70, side = "right"),
  #   Description = str_wrap(Description, width = 40),
  #   Description = fct_inorder(Description)
  # )

heat_terms <- top_terms %>%
  right_join(clsts, by = "Description") %>%
  arrange(cluster, desc(n), p.adjust) %>%
  group_by(cluster) %>%
  mutate(
    n_terms = n_distinct(Description),
    mean_p  = mean(p.adjust)
  ) %>%
  dplyr::slice(1:5) %>%
  arrange(cluster, p.adjust)

heat_dat <- go_new@compareClusterResult %>%
  filter(Description %in% heat_terms$Description) %>%
  left_join(clsts, by = "Description") %>%
  mutate(Description = fct_relevel(Description, rev(heat_terms$Description))) %>%
  arrange(Description) %>%
  mutate(
    Description = str_trunc(as.character(Description), width = 70, side = "right"),
    Description = str_wrap(Description, width = 40),
    Description = fct_inorder(Description)
  )
  
p_dat <- heat_dat %>%
  dplyr::filter(p.adjust < 0.05, qvalue < 0.2)

heat_dat %>%
  ggplot(aes(Cluster, Description, fill = as.character(cluster), alpha = -log10(p.adjust))) +
  geom_tile() +
  geom_point(
    aes(fill = NULL, alpha = NULL),
    data = p_dat,
    shape = 5,
    show.legend = FALSE
  ) +
  guides(fill = "none") +
  facet_wrap(
    ~ cluster, ncol = 1, scales = "free_y",
    labeller = as_labeller(c("1" = "Response to stimulus", "2" = "Response to virus", "3" = "Regulation of cytotoxicity"))) +
  scale_alpha_continuous(range = c(0, 1)
  ) +
  scale_fill_manual(values = c("lightblue", "orange", "red")) +
  djvdj_theme() +
  theme(
    axis.title = element_blank(),
    axis.text.x = element_text(angle = 45, hjust = 1)
  )

```

```{r "GO examples"}

# Example genes
early_gns <- top_degs %>%
  group_by(gene) %>%
  filter(n_distinct(subtype) > 2) %>%
  pull(gene) %>%
  unique()

plt_dat <- list("8dpi" = so_8, "24dpi" = so_24) %>%
  imap_dfr(~ {
    .x %>%
      FetchData(c("cell_type", "subtype", "orig.ident", "treatment", "rep", early_gns)) %>%
      as_tibble(rownames = ".cell_id") %>%
      mutate(tm = .y) %>%
      pivot_longer(all_of(early_gns))
  }) %>%
  filter(
    cell_type %in% c("Endothelial cells", "FRC", "Stromal cells", "Fibroblasts")
  ) %>%
  mutate(
    treatment = ifelse(treatment == "CHIKV", tm, str_c(treatment, "-", tm)),
    treatment = fct_relevel(treatment, c("mock-8dpi", "mock-24dpi", "8dpi", "24dpi")),
  )

typs <- unique(plt_dat$subtype)

typs %>%
  map(~ {
    plt_dat %>%
      ggplot(aes(treatment, value, alpha = rep, fill = treatment)) +
      geom_boxplot(
        outlier.size = 0.1, outlier.alpha = 1,
        position = position_dodge2(preserve = "single")
      ) +
      scale_alpha_manual(values = c(0.25, 0.5, 0.75)) +
      facet_wrap(~ name) +
      ggtitle(.x) +
      djvdj_theme() +
      theme(
        legend.position = "none",
        axis.title      = element_blank()
      )
  }) %>%
  plot_grid(plotlist = .)

```

```{r "hgWGCNA"}

library(hdWGCNA)

# Set input object
obj <- so_8 %>%
  mutate_meta(
    mutate,
    subtype = ifelse(lec_subtype != "other", lec_subtype, fib_subtype)
  )

typ_col <- "subtype"

# Set up object for hdWGCNA
obj <- obj %>%
  SetupForWGCNA(
    gene_select = "fraction",  # the gene selection approach
    fraction    = 0.01,        # fraction of cells that a gene needs to be expressed in order to be included
    wgcna_name  = "so_8"       # the name of the hdWGCNA experiment
  )

# construct metacells  in each group
obj <- obj %>%
  MetacellsByGroups(
    group.by    = c(typ_col, "treatment"),      # specify the columns in seurat_obj@meta.data to group by
    reduction   = "pca",                        # select the dimensionality reduction to perform KNN on
    k           = 20,                           # nearest-neighbors parameter
    # k           = 40,                           # nearest-neighbors parameter
    max_shared  = 10,                           # maximum number of shared cells between two metacells
    min_cells   = 40,                           # minimum number of cells to include group
    ident.group = typ_col                       # set the Idents of the metacell seurat object
)

# normalize metacell expression matrix:
obj <- NormalizeMetacells(obj)

# Set expression matrix for network analysis
obj <- obj %>%
  SetDatExpr(
    group_name = unlist(unique(obj@misc$so_8$wgcna_metacell_obj[[typ_col]]), use.names = FALSE),
    group.by   = typ_col,
    assay      = "RNA",
    slot       = "data"
  )

# Test different soft powers
# you can also use "unsigned" or "signed hybrid"
obj <- obj %>%
  TestSoftPowers(networkType = "signed")

# obj %>%
#   PlotSoftPowers() %>%
#   wrap_plots(ncol = 2)

# Construct co-expression network
# soft power is selected automatically
obj <- obj %>%
  ConstructNetwork(
    setDatExpr = FALSE,
    tom_name   = "so_8" # name of the topological overlap matrix written to disk
  )

# PlotDendrogram(obj)

# Compute module eigengenes
# need to run ScaleData first or else harmony throws an error
obj <- obj %>%
  ScaleData() %>%
  ModuleEigengenes(group.by.vars = "treatment")

# Compute eigengene-based connectivity (kME)
obj <- obj %>%
  ModuleConnectivity(
    group_name = unlist(unique(obj@misc$so_8$wgcna_metacell_obj[[typ_col]]), use.names = FALSE),
    group.by   = typ_col,
    harmonized = TRUE
  )

obj %>%
  qsave(file.path(params$so_dir, "so_8_wgcna.qs"))

# Plot genes ranked by kME for each module
# PlotKMEs(obj, ncol = 5)

modules <- GetModules(obj)

# Get hub genes for each module
hub_df <- GetHubGenes(obj, n_hubs = 100)

# compute gene scoring for the top 25 hub genes by kME for each module
# with Seurat method
obj <- obj %>%
  ModuleExprScore(
    n_genes = 100,
    method  = "Seurat"
  )

obj %>%
  ModuleFeaturePlot(
    features = "scores",  # plot the hub gene scores
    order    = "shuffle"  # order so cells are shuffled
  ) %>%
  wrap_plots(ncol = 6)

```

```{r "hgWGCNA DEGs"}

# Differentially expressed modules
bcs <- obj %>%
  colnames() %>%
  split(obj$treatment)

DMEs <- obj %>%
  FindDMEs(
    barcodes1  = bcs$mock,
    barcodes2  = bcs$CHIKV,
    test.use   = "wilcox"
    # group.by   = typ_col,
    # subset.ident = "fLEC"
    # wgcna_name ='INH'
  )

obj %>%
  PlotDMEsLollipop(
    DMEs,
    wgcna_name = "so_8",
    pvalue = "p_val_adj"
  )

# GO terms for each module
go_gns <- hub_df %>%
  pull(gene_name) %>%
  split(hub_df$module)

go_res <- go_gns %>%
  gost(organism = "mmusculus") %>%
  .$result %>%
  filter(term_size > 20, term_size < 1000, p_value < 0.05) %>%
  arrange(query, p_value)

# go_res %>%
#   filter(query %in% c("brown", "darkorange", "tan")) %>%
#   filter(query == "darkgreen")

# Heatmaps
mod_gns <- set_names(
  as.character(hub_df$module),
  hub_df$gene_name
)

heat_dat <- obj %>%
  FetchData(c(typ_col, "orig.ident", "treatment", "rep", names(mod_gns))) %>%
  pivot_longer(all_of(names(mod_gns))) %>%
  group_by(treatment, !!sym(typ_col), name) %>%
  summarize(value = mean(value), .groups = "drop")

plt_dat <- heat_dat %>%
  group_by(name) %>%
  mutate(
    value = as.numeric(scale(value)),
    mod   = mod_gns[name]
  ) %>%
  ungroup() %>%
  mutate(treatment = fct_relevel(treatment, c("mock", "CHIKV"))) %>%
  split(.$mod)

mods <- DMEs %>%
  filter(
    avg_log2FC < 0,
    module != "grey60"
  ) %>%
  pull(module) %>%
  head(5)
  
plts <- plt_dat[mods] %>%
  imap(~ {
    .x %>%
      ggplot(aes(treatment, name, fill = value)) +
      geom_tile() +
      facet_wrap(as.formula(str_c("~ ", typ_col)), scales = "free_y", nrow = 1) +
      scale_fill_gradientn(colours = c("lightblue", "white", "red")) +
      ggtitle(.y) +
      djvdj_theme() +
      theme(
        axis.text.y  = element_blank(),
        axis.ticks.y = element_blank(),
        axis.text.x  = element_text(angle = 45, hjust = 1),
        legend.position = "none"
      )
  })

plts %>%
  plot_grid(plotlist = ., align = "vh", axis = "trbl")

```

```{r "GSEA MARKERS", eval = FALSE}
# Cell types to test
# use all LEC and FRC subsets
# exclude CD45+ cell types and unassigned cells
cd45_typs <- c(
  "T cells", "B cells",
  "Macrophages", "unassigned"
)

typs <- so_24 %>%
  get_cell_types("subtype", "orig.ident", n_cells = 3)

typs <- typs[!typs %in% cd45_typs]

# Find marker genes for each cell type
marks_path <- here(params$so_dir, "all_subtype_markers.tsv")

if (!file.exists(marks_path)) {
  Idents(so_24) <- so_24 %>%
    FetchData("treatment")
  
  marks <- typs %>%
    map_dfr(~ {
      so_24 %>%
        subset(subset = subtype == .x) %>%
        find_conserved_markers(
          ident_1  = "CHIKV",
          grp_var  = "rep",
          p_max    = 1.1,
          fc_range = c(-Inf, Inf)
        ) %>%
        mutate(cell_type = .x)
    }) %>%
    arrange(desc(avg_log2FC))
  
  marks %>%
    write_tsv(marks_path)
  
} else {
  marks <- read_tsv(marks_path)
}

# Exclude ribosomal protein and mitochondrial genes
marks <- marks %>%
  filter(!grepl("^Rp[sl]|^mt-", gene))

```

```{r "H CLUSTERING", eval = FALSE}

# Marker stats columns
reps     <- unique(so_24$rep)
fc_clmns <- str_c(reps, "_avg_log2FC")
p_clmns  <- str_c(reps, "_p_val_adj")

# Format matrix
# Only include genes >1.25 fold change and <0.05 adj p-value
# This will cluster based only on fold changes for genes/cell types
# sinificantly upregulated, genes that are not upregulated are set as 0
sig_marks <- marks %>%
  filter(
    # if_all(all_of(fc_clmns), ~ .x > log2(1.25)),
    if_all(all_of(fc_clmns), ~ .x > log2(1)),
    if_all(all_of(p_clmns),  ~ .x < 0.05)
  ) %>%
  rowwise() %>%
  mutate(max_p = max(!!!syms(p_clmns))) %>%
  ungroup() %>%
  dplyr::select(gene, cell_type, avg_log2FC, max_p)
  
mat <- sig_marks %>%
  dplyr::select(-max_p) %>%
  pivot_wider(
    names_from  = cell_type,
    values_from = avg_log2FC,
    values_fill = 0
  ) %>%
  column_to_rownames("gene") %>%
  t() %>%
  scale() %>%
  t()

# Hierarchical clustering
clsts_path <- here(params$so_dir, "subtype_gene_clusters.tsv")

if (!file.exists(clsts_path)) {
  hcl <- mat %>%
    dist() %>%
    hclust(method = "ward.D")
  
  clsts <- cutree(hcl, k = 10)
  
  clsts <- tibble(
    gene    = names(clsts),
    cluster = as.character(unname(clsts))
  )
  
  clsts %>%
    write_tsv(clsts_path)
  
} else {
  clsts <- read_tsv(clsts_path, col_types = cols(cluster = col_character()))
}

# Summarize gene fold change for each cluster
# Calculate overall average fold change and p-value for each cell_type
# Use average fold change to identify the predominant cell types upregulated for
# each cluster
clst_dat <- marks %>%
  dplyr::select(gene, cell_type, avg_log2FC, max_pval) %>%
  right_join(clsts, by = "gene") %>%
  group_by(cluster, cell_type) %>%
  mutate(
    type_log2FC = mean(avg_log2FC),
    type_pval   = mean(max_pval),
  ) %>%
  ungroup()

# Identify cell types with upregulated genes for each cluster
# only include cell type if meets FC and p threshold for the cluster
clst_dat <- clst_dat %>%
  group_by(cluster) %>%
  mutate(
    max_fc     = max(type_log2FC),
    max_fc_lim = max(type_log2FC[type_log2FC < unique(max_fc)]) * 1.5
  ) %>%
  # filter(
  #   cell_type %in% cell_type[type_log2FC > log2(1.25) & type_pval < 0.05]
  # ) %>%
  ungroup()

# Further filter cell types for each cluster
# if one cell type is >1.5 fold higher than other cell types,
# remove other cell types
clst_dat <- clst_dat %>%
  group_by(cluster) %>%
  filter(
    n_distinct(cell_type) == 1 |
    cell_type %in% cell_type[
      (type_log2FC == max_fc & max_fc > max_fc_lim) |
      all(type_log2FC <= max_fc_lim)
    ]
  ) %>%
  mutate(
    up_types = list(unique(cell_type)),
    n_types = n_distinct(cell_type)
  ) %>%
  ungroup() %>%
  mutate(shared = n_types >= 5)

# Merge clusters with >4 positive cell types
# Create data.frame with new cluster IDs
# clst_dat <- clst_dat %>%
#   mutate(cluster = ifelse(shared, max(cluster[shared]), cluster))

```

```{r "H CLUSTERING HEATMAP", eval = FALSE}

# Identify top 6 clusters
n_top_clsts <- 100

up_types <- clst_dat %>%
  distinct(cluster, gene, up_types, n_types) %>%
  group_by(cluster, up_types, n_types) %>%
  summarize(n = n_distinct(gene), .groups = "drop")

top_clsts <- up_types %>%
  group_by(cluster, n_types, n) %>%
  summarize(id = map_chr(up_types, str_c, collapse = "/"), .groups = "drop") %>%
  # mutate(id = ifelse(n_types >= 5, "interferon response", id)) %>%
  group_by(cluster, id) %>%
  summarize(n = sum(n), .groups = "drop") %>%
  arrange(desc(n)) %>%
  mutate(
    rnk = row_number(),
    id  = str_replace(id, "_", " "),
    id  = str_c(id, "\n", n, " genes")
  ) 

clst_lvls <- set_names(top_clsts$id, top_clsts$cluster)

top_clsts <- head(top_clsts, n_top_clsts)
  
# Format scaled data for plotting
# Create new table with revised cluster IDs
# Want to show fold changes for all genes/cell types when creating heatmaps,
# not just upregulated genes
clsts <- clst_dat %>%
  distinct(gene, cluster) %>%
  filter(cluster %in% top_clsts$cluster)

plt_dat <- marks %>%
  filter(
    gene %in% unique(sig_marks$gene),
    cell_type %in% colnames(mat)
  ) %>%
  
  dplyr::select(gene, cell_type, avg_log2FC) %>%
  pivot_wider(
    names_from  = cell_type,
    values_from = avg_log2FC,
    values_fill = 0
  ) %>%
  column_to_rownames("gene") %>%
  t() %>%
  scale() %>%
  t() %>%

  as_tibble(rownames = "gene") %>%
  left_join(clsts, by = "gene") %>%
  filter(!is.na(cluster)) %>%
  pivot_longer(-c(gene, cluster), names_to = "cell_type") %>%
  mutate(
    cell_type = as.factor(cell_type),
    cluster = fct_relevel(cluster, top_clsts$cluster)
  ) %>%
  arrange(cluster, cell_type)
  
# ORIGINAL VERSION
# plt_dat <- mat %>%
#   as_tibble(rownames = "gene") %>%
#   left_join(clsts, by = "gene") %>%
#   filter(!is.na(cluster)) %>%  # NAs are genes not found in any cluster
#   pivot_longer(-c(gene, cluster), names_to = "cell_type") %>%
#   mutate(
#     cluster = fct_relevel(cluster, top_clsts$cluster)
#   ) %>%
#   arrange(cluster, cell_type)

# Format data for drawing heatmap boxes
type_lvls <- levels(plt_dat$cell_type)
type_lvls <- set_names(seq_along(type_lvls), type_lvls)

bx_dat <- up_types %>%
  filter(n_types < 3) %>%
  unnest(up_types) %>%
  group_by(cluster, up_types) %>%
  mutate(
    x = type_lvls[up_types],
    x = list(c(x - 0.5, x + 0.5))
  ) %>%
  ungroup() %>%
  unnest(x) %>%
  filter(cluster %in% plt_dat$cluster) %>%
  mutate(cluster = fct_relevel(cluster, levels(plt_dat$cluster)))

# Create heatmaps
clst_clrs <- c(
  "3" = lec_type_cols[["unassigned-LEC"]],
  "8" = lec_type_cols[["BEC"]],
  "5" = lec_type_cols[["PvC"]],
  "2" = lec_type_cols[["Macrophages"]],
  "4" = lec_type_cols[["Marco_LEC"]],
  "9" = lec_type_cols[["FRC"]]
)

h_heat <- plt_dat %>%
  # ggplot(aes(cell_type, gene, fill = value)) +
  # scale_fill_gradient2(low = "grey85", mid = "white", high = "#D7301F") +
  # scale_fill_gradient2(low = "#56B4E9", mid = "white", high = "#D7301F") +
  # guides(fill = guide_colorbar(
  #   ticks = FALSE,
  #   title.position = "bottom",
  #   barwidth = unit(150, "pt"), barheight = unit(10, "pt")
  # )) +
  
  ggplot(aes(cell_type, gene, fill = cluster, alpha = value)) +
  scale_alpha_continuous(range = c(0.01, 1)) +
  scale_fill_manual(values = clst_clrs) +
  guides(
    fill = "none",
    alpha = guide_legend(
      label.position = "bottom",
      keyheight = unit(7, "pt"),
      title = "z-score",
      title.position = "top"
    )
  ) +
  geom_vline(data = bx_dat, aes(xintercept = x)) +
  geom_tile() +
  facet_wrap(
    ~ cluster,
    scales         = "free_y",
    strip.position = "left",
    labeller       = as_labeller(clst_lvls),
    ncol = 4
  ) +
  scale_x_discrete(position = "top") +
  djvdj_theme() +
  theme(
    axis.text.y     = element_blank(),
    axis.ticks.y    = element_blank(),
    axis.text.x     = element_text(angle = 90, hjust = 0, vjust = 0.5),
    axis.title      = element_blank(),
    legend.position = "bottom",
    legend.title    = element_text(size = txt_pt1, hjust = 0.5),
    strip.clip      = "off"
  )

```

```{r "H CLUSTERING GO", eval = FALSE}

# Format genes for GO analysis
go_gns <- clsts %>%
  split(as.character(.$cluster)) %>%
  map(pull, gene)

# Identify GO terms
go_file <- file.path(params$so_dir, "subtype_gene_clusters_go.tsv")

if (!file.exists(go_file)) {
  go_res <- go_gns %>%
    gost(
      organism = "mmusculus",
      ordered_query = FALSE,
      evcodes = TRUE
    ) %>%
    .$result %>%
    filter(
      source %in% c("GO:BP", "GO:MF", "KEGG"),
      p_value < 0.05,
      term_size < 500
    ) %>%
    arrange(p_value)
  
  go_res %>%
    write_tsv(go_file)

} else {
  go_res <- read_tsv(go_file, col_types = cols(query = col_character()))
}

# Create GO bargraphs
lab_fn <- function(x) {
  x %>%
    str_remove("^[0-9]+-") %>%
    str_trunc(80) %>%
    str_wrap(45)
}

go_bars <- go_res %>%
  group_by(query) %>%
  dplyr::slice(1:6) %>%
  ungroup() %>%
  dplyr::select(-parents) %>%
  mutate(
    term_name = str_c(query, "-", term_name),
    term_name = fct_reorder(term_name, p_value, .desc = TRUE),
    query     = fct_relevel(query, names(clst_lvls))
  ) %>%
  ggplot(aes(-log10(p_value), term_name, fill = query)) +
  geom_col() +
  labs(x = "-log10(p-value)") +
  facet_wrap(~ query, scales = "free_y", strip.position = "right", ncol = 1) +
  scale_fill_manual(values = clst_clrs) +
  scale_y_discrete(labels = lab_fn) +
  djvdj_theme() +
  theme(
    axis.title.y = element_blank(),
    axis.title.x = element_text(size = txt_pt1),
    legend.position = "none"
    # strip.text = element_blank()
  )

```

```{r "H CLUSTERING TOP GENES", eval = FALSE}

# gns <- c("Cd74", "H2-Ab1", "Lgmn", "Angpt2", "Pik3r1", "Gja1", "Ywhag")    # BEC
# gns <- c("Vcam1", "Akt3", "Cdh5", "Ccl5", "F11r", "Olr1", "Dst", "Nlgn2")  # Marco
# gns <- c("Cxcl9", "Cxcl10", "Ccl2", "Ccl7", "Ccl4", "Il1b", "Il6", "Tnf")  # All
# gns <- c("Nod1", "Tlr3")                                                   # FRC/PvC
# gns <- c("Oas2", "Oas3", "Fgl2", "Irf1", "Cxcl10")                         # Mac

gns <- c("Irf7", "Ccl2", "Cxcl9", "Oas3", "Vcam1", "Tlr3")

# Format plotting data
dat <- so_24 %>%
  FetchData(c("treatment", "orig.ident", "lec_type", gns)) %>%
  rownames_to_column("cell_id") %>%
  pivot_longer(all_of(gns), names_to = "gene") %>%
  filter(lec_type %in% colnames(mat)) %>%
  mutate(
    treatment = fct_relevel(treatment, treats),
    gene = fct_relevel(gene, gns)
  )
  
# Format data for p-value labels
p_dat <- sig_marks %>%
  mutate(p_lab = format_pvalue(max_p)) %>%
  dplyr::rename(lec_type = cell_type) %>%
  filter(gene %in% gns) %>%
  mutate(gene = fct_relevel(gene, gns))

# Create boxplots
bx_fn <- function(x) str_c(x, "\nexpression")

gn_bxs <- dat %>%
  ggplot(aes(lec_type, value, alpha = treatment, fill = lec_type, color = lec_type)) +
  
  geom_boxplot(
    outlier.size = 0.001,
    position = position_dodge2(preserve = "single"),
    key_glyph = draw_key_point
  ) +

  geom_richtext(
    aes(lec_type, Inf, label = p_lab),
    data = p_dat,
    alpha = 1,
    size  = 5 / .pt,
    color = "black",
    fill  = NA,
    label.color = NA,
    vjust = "inward"
  ) +
    
  facet_wrap(
    ~ gene,
    scales = "free_y",
    ncol = 1,
    strip.position = "left",
    labeller = as_labeller(bx_fn)
  ) +
  guides(
    fill = "none",
    color = "none",
    alpha = guide_legend(override.aes = list(shape = 15, alpha = c(0.5, 1), size = 4))
  ) +
  scale_y_continuous(expand = expansion(c(0.05, 0.2))) +
  scale_alpha_manual(values = c(0.1, 0.6)) +
  scale_fill_manual(values = lec_type_cols) +
  scale_color_manual(values = lec_type_cols) +
  djvdj_theme() +
  theme(
    axis.title      = element_blank(),
    axis.text.x     = element_text(angle = 45, hjust = 1),
    legend.title    = element_blank(),
    legend.position = "top",
    strip.placement = "outside"
    # plot.margin = margin(15, 15, 15, 30),
  )

```

```{r "H FIGURE", fig.height = 10, fig.width = 10, eval = FALSE}

plot_grid(
  h_heat, go_bars, gn_bxs,
  align = "h",
  axis  = "tb",
  nrow  = 1,
  rel_widths = c(0.75, 1, 1),
  labels = c("A", "B", "C"),
  label_size = ttl_pt2 * 1.5
)

```

```{r "H SUPP", eval = FALSE}

# Format cluster labels
tab_labs <- clst_lvls %>%
  map_chr(str_remove, "\n.+") %>%
  map_chr(str_replace, "[/ ]", "_")

# Format table for gene clusters
tab_dat <- clst_dat %>%
  group_by(gene, cluster) %>%
  summarize(
    avg_log2FC = mean(avg_log2FC),
    # pval       = mean(max_pval),
    .groups    = "drop"
  ) %>%
  mutate(cluster = tab_labs[cluster]) %>%
  group_by(cluster) %>%
  mutate(n_genes = n_distinct(gene)) %>%
  ungroup() %>%
  arrange(desc(n_genes), desc(avg_log2FC)) %>%
  as.data.frame()

# Write excel file
tab_labs %>%
  walk(~ {
    tab_dat %>%
      filter(cluster == .x) %>%
      xlsx::write.xlsx(
        file      = file.path(params$so_dir, "gene_clusters.xlsx"),
        sheetName = .x,
        append    = TRUE,
        col.names = TRUE,
        row.names = FALSE
      )
  })

```

```{r "CLUSTERING IFN GENES"}

# Cluster IFN genes
# EXCLUDE CD45+ CELL TYPES
cd45_typs <- c("T cells", "B cells", "Macrophages", "unassigned")

ifn_clst <- top_clsts %>%
  filter(grepl("interferon", id)) %>% 
  pull(cluster)

ifn_gns <- clsts %>%
  filter(cluster == ifn_clst) %>%
  pull(gene)

ifn_mat <- mat[ifn_gns, !colnames(mat) %in% cd45_typs]

hcl <- ifn_mat %>%
  dist() %>%
  hclust(method = "ward.D")

ifn_clsts <- cutree(hcl, k = 10)

ifn_clsts <- tibble(
  gene    = names(ifn_clsts),
  cluster = as.character(unname(ifn_clsts))
)

# Format data for heatmaps
# Summarize gene fold change for each cluster
# Calculate overall average fold change and p-value for each cell_type
# Use average fold change to identify the predominant cell types upregulated for
# each cluster
clst_dat <- marks %>%
  dplyr::select(gene, cell_type, avg_log2FC, max_pval) %>%
  right_join(ifn_clsts, by = "gene") %>%
  group_by(cluster, cell_type) %>%
  mutate(
    type_log2FC = mean(avg_log2FC),
    type_pval   = mean(max_pval),
  ) %>%
  ungroup()

# Only include cell type if meets FC and p threshold for the cluster
# EXCLUDE CD45+ CELL TYPES
clst_dat <- clst_dat %>%
  group_by(cluster) %>%
  mutate(
    max_fc     = max(type_log2FC),
    max_fc_lim = max(type_log2FC[type_log2FC < unique(max_fc)]) * 1.5
  ) %>%
  filter(
    cell_type %in% cell_type[type_log2FC > log2(1.25) & type_pval < 0.05],
    !cell_type %in% cd45_typs
  )

# Further filter cell types for each cluster
# if one cell type is >1.5 fold higher than other cell types,
# remove other cell types
clst_dat <- clst_dat %>%
  filter(
    n_distinct(cell_type) == 1 |
    cell_type %in% cell_type[
      (type_log2FC == max_fc & max_fc > max_fc_lim) |
      all(type_log2FC <= max_fc_lim)
    ]
  ) %>%
  mutate(
    up_types = list(unique(cell_type)),
    n_types = n_distinct(cell_type)
  ) %>%
  ungroup() %>%
  mutate(shared = n_types >= 5)

# Merge clusters with >4 positive cell types
# Create data.frame with new cluster IDs
# clst_dat <- clst_dat %>%
#   mutate(cluster = ifelse(shared, max(cluster[shared]), cluster))

```

```{r "IFN HEATMAPS"}

# Identify top 6 clusters
up_types <- clst_dat %>%
  distinct(cluster, gene, up_types, n_types) %>%
  group_by(cluster, up_types, n_types) %>%
  summarize(n = n_distinct(gene), .groups = "drop")

top_clsts <- up_types %>%
  group_by(cluster, n_types, n) %>%
  summarize(id = map_chr(up_types, str_c, collapse = "/"), .groups = "drop") %>%
  mutate(id = ifelse(n_types >= 5, "interferon response", id)) %>%
  group_by(cluster, id) %>%
  summarize(n = sum(n), .groups = "drop") %>%
  arrange(desc(n)) %>%
  mutate(
    rnk = row_number(),
    id  = str_replace(id, "_", " "),
    id = str_c(id, "\n", n, " genes")
  ) 

clst_lvls <- set_names(top_clsts$id, top_clsts$cluster)

top_clsts <- head(top_clsts, 6)
  
# Format scaled data for plotting
# Create new table with revised cluster IDs
# Want to show fold changes for all genes/cell types when creating heatmaps,
# not just upregulated genes
clsts <- clst_dat %>%
  distinct(gene, cluster) %>%
  filter(cluster %in% top_clsts$cluster)

plt_dat <- marks %>%
  filter(
    gene %in% unique(sig_marks$gene),
    cell_type %in% colnames(ifn_mat)
  ) %>%
  
  dplyr::select(gene, cell_type, avg_log2FC) %>%
  pivot_wider(
    names_from  = cell_type,
    values_from = avg_log2FC,
    values_fill = 0
  ) %>%
  column_to_rownames("gene") %>%
  t() %>%
  scale() %>%
  t() %>%

  as_tibble(rownames = "gene") %>%
  left_join(clsts, by = "gene") %>%
  filter(!is.na(cluster)) %>%
  pivot_longer(-c(gene, cluster), names_to = "cell_type") %>%
  mutate(
    cell_type = as.factor(cell_type),
    cluster = fct_relevel(cluster, top_clsts$cluster)
  ) %>%
  arrange(cluster, cell_type)

# Format data for drawing heatmap boxes
type_lvls <- levels(plt_dat$cell_type)
type_lvls <- set_names(seq_along(type_lvls), type_lvls)

bx_dat <- up_types %>%
  filter(n_types < 3) %>%
  unnest(up_types) %>%
  group_by(cluster, up_types) %>%
  mutate(
    x = type_lvls[up_types],
    x = list(c(x - 0.5, x + 0.5))
  ) %>%
  ungroup() %>%
  unnest(x) %>%
  filter(cluster %in% plt_dat$cluster) %>%
  mutate(cluster = fct_relevel(cluster, levels(plt_dat$cluster)))

# Create heatmaps
clst_clrs <- c(
  "3" = lec_type_cols[["unassigned-LEC"]],
  "8" = lec_type_cols[["BEC"]],
  "5" = lec_type_cols[["PvC"]],
  "2" = lec_type_cols[["Macrophages"]],
  "4" = lec_type_cols[["Marco_LEC"]],
  "9" = lec_type_cols[["FRC"]]
)

h_heat <- plt_dat %>%
  ggplot(aes(cell_type, gene, fill = cluster, alpha = value)) +
  scale_alpha_continuous(range = c(0.01, 1)) +
  scale_fill_manual(values = clst_clrs) +
  guides(
    fill = "none",
    alpha = guide_legend(
      label.position = "bottom",
      keyheight = unit(7, "pt"),
      title = "z-score",
      title.position = "top"
    )
  ) +
  geom_vline(data = bx_dat, aes(xintercept = x)) +
  geom_tile() +
  facet_wrap(
    ~ cluster,
    scales         = "free_y",
    strip.position = "left",
    labeller       = as_labeller(clst_lvls),
    ncol = 1
  ) +
  scale_x_discrete(position = "top") +
  djvdj_theme() +
  theme(
    axis.text.y  = element_blank(),
    axis.ticks.y = element_blank(),
    axis.text.x  = element_text(angle = 90, hjust = 0, vjust = 0.5),
    axis.title   = element_blank(),
    legend.position = "bottom",
    legend.title    = element_text(size = txt_pt1, hjust = 0.5),
    strip.clip = "off"
  )

```

