---
title: "CHIKV LNSC figures"
author: "Ryan Sheridan"
date: "`r Sys.Date()`"
output: html_document

params:
  template_dir: "src"
  so_dir:       "~/Dropbox/Ryan/Projects/morrison-lnsc/results/sobjs"
  orig_so_dir:  "~/Dropbox/Ryan/Projects/morrison-scRNA-seq/results"
---

<br>

```{r "chunk opts", echo = FALSE}
# Default chunk options
knitr::opts_chunk$set(
  message = FALSE,
  warning = FALSE,
  echo    = FALSE,
  dpi     = 500
)
```

```{r "packages"}
# tidyverse
library(tidyverse)
library(here)
library(knitr)
library(broom)

# plotting
library(scales)
library(ggrepel)
library(ggtrace)
library(patchwork)
library(cowplot)
library(colorblindr)
library(ggtext)

# analysis
library(Seurat)
library(clustifyr)
library(clustifyrdata)
library(DoubletFinder)
library(M3Drop)
library(CellChat)
library(djvdj)
library(presto)
library(gprofiler2)
library(qs)

library(clusterProfiler)
library(enrichplot)
library(msigdbr)

source(here(params$template_dir, "funs.R"))

chikv_grps <- c("CHIKV-low", "CHIKV-high")
```

```{r "functions"}
type_lab_fn <- function(x) str_replace(x, "_", " ")

plot_example_gene <- function(so_in, feat, grp, box_colors = NULL,
                              umap_colors = c("#56B4E9", "white", "#D7301F")) {
  
  dat <- so_in %>%
    FetchData(c(feat, grp, "UMAP_1", "UMAP_2"))
  
  u <- dat %>%
    arrange(!!sym(feat)) %>%
    ggplot(aes(UMAP_1, UMAP_2, fill = !!sym(feat))) +
    geom_point_trace(size = 0.25, stroke = 0.65) +
    scale_fill_gradientn(colours = umap_colors) +
    guides(fill = guide_colorbar(
      ticks          = FALSE,
      barheight      = unit(10, "pt"),
      barwidth       = unit(140, "pt"),
      title.position = "top"
    )) +
    umap_theme +
    theme(
      legend.position      = "top",
      legend.title         = element_text(hjust = 0.5),
      legend.justification = 0.5,
      panel.border         = element_rect(fill = NA, color = "grey85")
    )
  
  
  lvls <- dat %>%
    group_by(!!sym(grp)) %>%
    summarize(
      med   = median(!!sym(feat)),
      stats = list(boxplot.stats(!!sym(feat))),
      q3    = map_dbl(stats, ~ .x$stats[4]),
      q4    = map_dbl(stats, ~ .x$stats[5]),
      max   = max(!!sym(feat)),
      .groups = "drop"
    ) %>%
    arrange(med, q3, q4, max) %>%
    pull(grp) %>%
    rev()
  
  b <- dat %>%
    plot_violin(
      feat, grp,
      plot_colors   = box_colors,
      plot_lvls     = lvls,
      method        = "boxplot",
      outlier.size  = 0.25,
      outlier.alpha = 1
    ) +
    theme(
      legend.position = "none",
      aspect.ratio = 0.21,
      axis.text.x  = element_text(angle = 45, hjust = 1),
      axis.line.y  = element_line(size = 0.5, color = "grey85"),
      axis.ticks.x = element_blank(),
      panel.border = element_blank(),
      axis.title.x = element_blank()
    )
  
  plot_grid(
    u, b,
    ncol        = 1,
    rel_heights = c(1, 0.5),
    align       = "v",
    axis        = "rl"
  )
}
```

```{r "load objects"}
# Replace this with setup.Rmd that has code used to process each sample
so_8  <- read_rds(file.path(params$orig_so_dir, "2021-07-16/sobjs/so.rds"))
so_24 <- read_rds(file.path(params$orig_so_dir, "2021-01-08/sobjs/so.rds"))

so_lec_8  <- read_rds(file.path(params$orig_so_dir, "2021-07-16/sobjs/so_lec.rds"))
so_lec_24 <- read_rds(file.path(params$orig_so_dir, "2021-01-08/sobjs/so_lec.rds"))

# Adjust 24 hpi
# Add PvCs to cell types
# This is based on clustifyr results using ref_lymphnodestromal, and Pecam1,
# Pdpn, Itga7, and Acta2 expression
lec_cell_types <- "Endothelial cells"
fib_cell_types <- c("Fibroblasts", "Stromal cells")

so_24 <- so_24 %>%
  mutate_meta(
    mutate,
    cell_type   = case_when(
      fib_type == "PvC"             ~ fib_type,
      cell_type %in% fib_cell_types ~ "FRC",
      TRUE                          ~ cell_type
    )
  )

so_24 <- so_24 %>%
  mutate_meta(
    mutate,
    lec_type = case_when(
      lec_subtype == "unassigned"   ~ "unassigned-LEC",
      cell_type %in% lec_cell_types ~ lec_subtype,
      TRUE                          ~ cell_type
    )
  )

so_24 <- so_24 %>%
  mutate_meta(
    mutate,
    chikv_type = if_else(chikv_grp == chikv_grps[2], lec_type, chikv_grp)
  )

# Adjust 8 hpi
# Cluster 27 appears to be FRCs based on Pdpn, Pecam1, and Pdgfra
so_8 <- so_8 %>%
  mutate_meta(
    mutate, 
    orig.ident = str_c(treatment, "-", rep),
    lec_type = recode(
      lec_type,
      `unassigned-LEC`     = "unassigned",
      Fibroblasts          = "FRC",
      `Stromal cells (DN)` = "PvC"
    ),
    chikv_grp = ifelse(tot_nCount_CHIKV > 0, chikv_grps[2], chikv_grps[1])
  )

so_8 <- so_8 %>%
  mutate_meta(
    mutate,
    lec_type = ifelse(RNA_snn_res.5 == "27", "FRC", lec_type)
  )

# Merge objects
so_merge_df <- bind_rows(so_8@meta.data, so_24@meta.data)

# Set levels
sam_lvls   <- unique(so_24$orig.ident)
treats     <- unique(so_24$treatment)
```

```{r "references"}
# LEC reference
ref_lec <- ref_immgen[, grepl("^Endothelial cells", colnames(ref_immgen))]
ref_lec <- ref_lec[rownames(ref_lec) %in% rownames(ref_LEC_xiang), ]

colnames(ref_lec) <- colnames(ref_lec) %>%
  str_replace("Endothelial cells \\(BEC\\)", "BEC")

ref_LEC_xiang <- ref_LEC_xiang[rownames(ref_lec), ]

if (!identical(rownames(ref_LEC_xiang), rownames(ref_lec))) {
  stop("LEC reference rownames do not match.")
}

ref_lec <- cbind(ref_LEC_xiang, ref_lec)
ref_lec <- ref_lec[, !grepl("^Endothelial cells", colnames(ref_lec))]

# Fibroblast/stromal cell reference
ref_fib <- ref_immgen[, grepl("^Fibroblast", colnames(ref_immgen))]
ref_fib <- ref_fib[rownames(ref_fib) %in% rownames(ref_lymphnodestromal), ]

ref_lymphnodestromal <- ref_lymphnodestromal[rownames(ref_fib), ]

if (!identical(rownames(ref_lymphnodestromal), rownames(ref_fib))) {
  stop("Fibroblast/stromal reference rownames do not match.")
}

ref_fib <- cbind(ref_lymphnodestromal, ref_fib)
```

```{r "theme"}
# ggplot2 themes
txt_pt1  <- 10
txt_pt2  <- 12
ttl_pt1  <- 14
ttl_pt2  <- 16
ln_pt    <- 0.5
ln_col   <- "grey85"
cir_size <- 3.5
sqr_size <- 4

txt_theme_1 <- theme(
  strip.text  = element_text(size = ttl_pt2),
  legend.text = element_text(size = txt_pt2),
  axis.title  = element_text(size = ttl_pt2),
  axis.text   = element_text(size = txt_pt2)
)

line_theme <- theme(
  axis.line.x  = element_line(linewidth = ln_pt, color = ln_col),
  axis.line.y  = element_line(linewidth = ln_pt, color = ln_col),
  axis.ticks.x = element_line(size = ln_pt, color = ln_col),
  axis.ticks.y = element_line(size = ln_pt, color = ln_col)
)

base_theme <- djvdj_theme() +
  theme(
    plot.title       = element_text(face = "plain", size = ttl_pt2),
    strip.background = element_blank(),
    strip.text       = element_text(face = "plain")
  ) +
  txt_theme_1

umap_theme <- base_theme +
  theme(
    panel.border = element_rect(color = NA, fill = NA),
    axis.title   = element_blank(),
    axis.line.x  = element_blank(),
    axis.line.y  = element_blank(),
    axis.ticks.x = element_blank(),
    axis.ticks.y = element_blank(),
    axis.text    = element_blank()
  )

fade_0 <- "#FAFAFA"
fade_1 <- "#F0F0F0"
fade_2 <- "#D9D9D9"

hist_y_lab <- "number of cells"

# alpha for plots
al <- 0.7

# Okabe Ito color palettes
ito_cols <- c(
  palette_OkabeIto[c(1:3, 5)], "#D7301F",
  palette_OkabeIto[c(4, 6)],   "#6A51A3",
  palette_OkabeIto[7],         "#875C04",
  "#065D43", "#F7AF34", "#00446E"
)

ito_cols <- ito_cols[3:length(ito_cols)] %>%
  darken(0.2) %>%
  c(ito_cols, ., "#686868", "#000000")

# Set sample colors
get_cols <- create_col_fun(ito_cols)

sam_cols <- c(
  # "#F03B20", "#FD8D3C",
  "#BD0026", "#FECC5C", "#FFFFB2",
  # "#0072B2", "#56B4E9",
  "#00446E", "#035B8F", "#2A8FBF"
)

names(sam_cols) <- sam_lvls

# CHIKV treatment groups
chikv_infctd <- so_merge_df %>%
  group_by(treatment) %>%
  summarize(mn = mean(nCount_CHIKV), .groups = "drop") %>%
  filter(mn == max(mn)) %>%
  pull(treatment)

# Cell type colors
type_cols <- unique(so_merge_df$cell_type)
type_cols <- set_names(ito_cols[seq_along(type_cols)], type_cols)

type_cols["unassigned"] <- "#999999"

subtype_cols <- c(colnames(ref_lec), colnames(ref_fib))
subtype_cols <- set_names(ito_cols[seq_along(subtype_cols)], subtype_cols)
subtype_cols["unassigned"] <- "#999999"
subtype_cols["other"]      <- fade_0

lec_type_cols <- c(subtype_cols, type_cols[names(type_cols) != "unassigned"])
lec_type_cols["CHIKV-low"]      <- "white"
lec_type_cols["T cells"]        <- "black"
lec_type_cols["B cells"]        <- "#875C04"
lec_type_cols["FRC"]            <- "#00375B"
lec_type_cols["PvC"]            <- "#065D43"
lec_type_cols["unassigned-LEC"] <- "#676767"


lec_type_cols <- so_merge_df %>%
  pull(lec_type) %>%
  table() %>%
  sort(decreasing = TRUE) %>%
  names()

lec_type_cols <- lec_type_cols[lec_type_cols != "unassigned"]

lec_type_cols <- set_names(
  ito_cols[seq_along(lec_type_cols)],
  lec_type_cols
)

lec_type_cols["unassigned"] <- "grey70"

old_cols <- lec_type_cols

lec_type_cols["FRC"]   <- old_cols["Valve"]
lec_type_cols["Valve"] <- old_cols["fLEC"]
lec_type_cols["fLEC"]  <- old_cols["FRC"]

# CHIKV clusters colors
grp_cols <- set_names(
  c("#56B4E9", "#0072B2"),
  chikv_grps
)

n_reps <- n_distinct(so_merge_df$rep)

grp_rep_cols <- grp_cols %>% 
  imap(~ set_names(
    rep(.x, n_reps),
    str_c(.y, "-", 1:n_reps)
  )) %>%
  flatten_chr()
```

## Figure 1

Chikungunya virus is found in MARCO-expressing floor lymphatic endothelial cells
in the draining lymph node at 8 hpi

```{r "fig 1A", fig.width = 10, fig.height = 10}
# Format plot data
dat <- so_8@meta.data %>%
  arrange(tot_pct_CHIKV)

# Create UMAP
pseudo <- dat$tot_pct_CHIKV
pseudo <- min(pseudo[pseudo > 0]) / 2

A <- dat %>%
  ggplot(aes(UMAP_1, UMAP_2, fill = tot_pct_CHIKV + pseudo)) +
  geom_point_trace(
    color = "grey",
    trace_position = "bottom"
  ) +
  scale_fill_gradientn(colours = c("white", "#D7301F"), trans = "log10", labels = scales::label_number()) +
  guides(fill = guide_colorbar(title.position = "top", ticks = FALSE, title = "% CHIKV counts")) +
  umap_theme +
  theme(
    aspect.ratio         = 0.9,
    legend.position      = "top",
    legend.title         = element_text(hjust = 0.5, size = ttl_pt2 * 1.5),
    legend.text          = element_text(angle = 45, hjust = 1), 
    legend.direction     = "horizontal",
    legend.justification = 0.2,
    legend.key.width     = unit(30, "pt"),
    legend.key.height    = unit(5, "pt")
  )
```

```{r "fig 1C", fig.width = 10, fig.height = 10}
# Format plot data
dat <- so_8@meta.data %>%
  filter(chikv_grp == chikv_grps[2])

plt_labs <- get_nlab_fun(dat, "lec_type", l = "\n", r = "")
plt_labs <- plt_labs()
plt_labs <- map_chr(plt_labs, type_lab_fn)

dat <- dat %>%
  mutate(lec_type = fct_relevel(lec_type, names(plt_labs)))

# Create UMAP
C <- so_8@meta.data %>%
  mutate(lec_type = fct_relevel(lec_type, names(plt_labs))) %>%
  ggplot(aes(UMAP_1, UMAP_2)) +
  geom_point_trace(
    color = "grey",
    fill  = "white",
    trace_position = "bottom"
  ) +
  geom_point(aes(color = lec_type), data = dat) +
  scale_color_manual(values = lec_type_cols, labels = plt_labs) +
  scale_y_discrete(labels = type_lab_fn) +
  guides(color = guide_legend(override.aes = list(size = 4))) +
  labs(title = "CHIKV-high cell types") +
  umap_theme +
  theme(
    aspect.ratio      = 0.9,
    plot.title        = element_text(size = ttl_pt2 * 1.5, hjust = 0.2),
    legend.title      = element_blank(),
    legend.key.height = unit(30, "pt"),
    legend.text       = element_text(size = ttl_pt1)
  )
```

```{r "fig 1B", fig.width = 3, fig.height = 6}
# Format plot data
# ONLY PLOTTING ENRICHED COUNTS
dat <- so_8@meta.data %>%
  filter(treatment == "CHIKV") %>%
  group_by(lec_type) %>%
  filter(n() > 50) %>%
  ungroup() %>%
  mutate(
    counts = tot_nCount_CHIKV - nCount_CHIKV
    # counts = tot_nCount_CHIKV
    # enr = tot_nCount_CHIKV + 1
  )

# Set plot levels
lvls <- dat %>%
  group_by(lec_type) %>%
  summarize(
    med   = median(counts),
    stats = list(boxplot.stats(counts)),
    q3    = map_dbl(stats, ~ .x$stats[4]),
    q4    = map_dbl(stats, ~ .x$stats[5]),
    max   = max(counts),
    .groups = "drop"
  ) %>%
  arrange(med, q3, q4, max) %>%
  pull(lec_type)

plt_labs <- get_nlab_fun(dat, "lec_type", l = "\n", r = "")
plt_labs <- plt_labs()
plt_labs <- map_chr(plt_labs, type_lab_fn)

# Create boxplots
B <- dat %>%
  mutate(lec_type = fct_relevel(lec_type, lvls)) %>%
  ggplot(
    aes(counts + 1, lec_type, fill = lec_type, color = lec_type, alpha = rep)
  ) +
  geom_boxplot(size = 1, outlier.size = 1.5, outlier.alpha = 1) +
  scale_x_log10() +
  scale_y_discrete(labels = plt_labs)+
  scale_fill_manual(values = lec_type_cols) +
  scale_color_manual(values = lec_type_cols) +
  scale_alpha_manual(values = c(0.5, 0.5)) +
  labs(x = "CHIKV counts + 1") +
  base_theme +
  theme(
    legend.position = "none",
    plot.margin     = margin(100, 5, 5, 5),
    axis.title.y    = element_blank(),
    axis.title.x    = element_text(size = ttl_pt2),
    axis.text.y     = element_text(size = ttl_pt2)
  )
```

```{r "fig 1D", fig.width = 3, fig.height = 5}
# Format plot data
dat <- so_8 %>%
  FetchData(c("lec_type", "treatment", "rep", "orig.ident", "chikv_grp", "Marco")) %>%
  filter(lec_type == "fLEC") %>%
  mutate(chikv_grp = ifelse(treatment == "mock", treatment, chikv_grp))

clrs <- set_names(
  c("grey", rep(lec_type_cols[["fLEC"]], 2)),
  c("mock", chikv_grps)
)

# Calculate p-values
p_vals <- dat %>%
  calc_p_vals(
    data_column = "Marco",
    type_column = "chikv_grp"
  )

comps <- names(clrs) %>%
  combn(2, simplify = FALSE)

comps <- list(
  x    = map_chr(comps, pluck, 1),
  xend = map_chr(comps, pluck, 2),
  y    = c(5, 5.6, 5)
)

# Create boxplots
D <- dat %>%
  plot_violin(
    "Marco", "chikv_grp",
    method        = "boxplot",
    plot_colors   = clrs,
    alpha         = c(0.5, 0.1, 0.5),
    size          = 1.5,
    outlier.size  = 1.5,
    outlier.alpha = 1,
    width         = 0.6
  ) +
  labs(title = "fLEC", y = "Marco expression") +
  base_theme +
  theme(
    aspect.ratio = 1.5,
    plot.title   = element_text(size = ttl_pt2 * 1.5),
    axis.title.y = element_text(size = ttl_pt2),
    axis.text.x  = element_text(size = ttl_pt2)
  )

# Add p-values to plot
comps %>%
  pwalk(~ {
    v_args <- list(...)[c("x", "xend")]
    
    p <- p_vals %>%
      filter(`Cell type 1` %in% v_args & `Cell type 2` %in% v_args) %>%
      pull(p_adj)
    
    D <<- D %>%
      add_pvals(
        ...,
        p_val    = p,
        size     = txt_pt2 / .pt,
        line_col = "grey75"
      )
  })
```

```{r "fig 1", fig.width = 15, fig.height = 20}
l <- plot_grid(
  A, C,
  labels = c("A", "C"), label_size = ttl_pt2 * 1.5,
  ncol   = 1,
  align  = "v",
  axis   = "rl"
)

r <- plot_grid(
  B, D,
  labels = c("B", "D"), label_size = ttl_pt2 * 1.5,
  ncol   = 1,
  align  = "v",
  axis   = "rl"
)

plot_grid(l, r, nrow = 1, rel_widths = c(1, 0.5))
```

<br>

## Figure S1

CHIKV may target MARCO-expressing LECs for infection

```{r "fig S1A", fig.width = 10, fig.height = 10}
# Pseudo count for pct_mito
pseudo <- so_24$pct_mito
pseudo <- min(pseudo[pseudo > 0]) / 2

# Format data
dat <- so_24@meta.data %>%
  as_tibble(rownames = "cell_id") %>%
  mutate(pct_mito = pct_mito + pseudo) %>%
  arrange(CHIKV_sgRNA_fc)

# sgRNA ratio UMAP
A <- dat %>%
  ggplot(aes(UMAP_1, UMAP_2, fill = CHIKV_sgRNA_fc)) +
  geom_point_trace(
    color = "grey",
    trace_position = "bottom",
    size = 0.5
  ) +
  scale_fill_gradientn(colours = c("white", "#0072B2", "#002037"), labels = scales::label_number()) +
  guides(fill = guide_colorbar(title.position = "top", ticks = FALSE, title = "CHIKV sgRNA ratio")) +
  umap_theme +
  theme(
    aspect.ratio         = 0.9,
    legend.position      = "top",
    legend.title         = element_text(hjust = 0.5, size = ttl_pt2),
    legend.direction     = "horizontal",
    legend.justification = 0.5,
    legend.key.width     = unit(30, "pt"),
    legend.key.height    = unit(5, "pt")
  )
```

```{r "fig S1B", fig.width = 10, fig.height = 7}
# Scatter plots
y_vars <- c(
  "# mouse genes\nexpressed" = "nFeature_RNA",
  "% mitochondrial\nreads" = "pct_mito"
)

typs <- c("Marco_LEC", "PvC", "BEC")

dat <- dat %>%
  filter(
    chikv_grp == chikv_grps[2],
    lec_type %in% typs
  ) %>%
  mutate(lec_type = fct_relevel(lec_type, typs)) %>%
  dplyr::rename(!!!syms(y_vars)) %>%
  pivot_longer(all_of(names(y_vars))) %>%
  group_by(lec_type, name) %>%
  mutate(
    cor = cor(log10(CHIKV_sgRNA_fc), log10(value)),
    cor = str_c("r = ", round(cor, 2))
  ) %>%
  ungroup()

B <- dat %>%
  split(.$lec_type) %>%
  unname() %>%
  imap(~ {
    clmn <- .y
    
    .x <- .x %>%
      split(.$name)
    
    top <- names(.x)[1]
    
    .x %>%
      imap(~ {
        if (clmn == 1) y_ttl <- .y
        else           y_ttl <- NULL
        
        ttl <- NULL
        sub <- NULL
        x_ttl <- NULL
        
        if (.y == top) {
          ttl   <- type_lab_fn(unique(.x$lec_type))
          sub   <- str_c("n = ", nrow(.x))
        } else if (clmn == 2) {
          x_ttl <- "CHIKV sgRNA ratio"
        }
        
        .x %>%
          ggplot(aes(CHIKV_sgRNA_fc, value, color = lec_type)) +
          geom_point() +
          geom_smooth(
            method = "lm", formula = y ~ x,
            color = "black", linewidth = 0.5, linetype = 2, alpha = 0.2) +
          geom_text(
            aes(Inf, Inf, label = cor),
            hjust = 1.2, vjust = 1.2,
            color = "black",
            check_overlap = TRUE
          ) +
          scale_color_manual(values = lec_type_cols) +
          scale_x_log10() +
          scale_y_log10(expand = expansion(c(0.05, 0.10))) +
          labs(title = ttl, subtitle = sub, x = x_ttl, y = y_ttl) +
          theme(aspect.ratio = 0.9) +
          base_theme +
          theme(legend.position = "none")
      })
  }) %>%
  flatten() %>%
  wrap_plots(byrow = FALSE, nrow = 2)
```

```{r "fig S1C", fig.width = 10, fig.height = 6.5}
# Genes to plot
gns <- c(
  "Lyve1", "Marco",    # Marco LECs
  "Madcam1", "Vcam1",  # fLECs
  "Ptx3",              # Ptx3 LECs
  "Ackr4"              # cLECs
)

# Format data
dat <- list("8hpi" = so_8, "24hpi" = so_24) %>%
  imap_dfr(~ {
    .x %>%
      FetchData(c("lec_type", "treatment", "rep", gns)) %>%
      as_tibble(rownames = "cell_id") %>%
      mutate(tm = .y)
  }) %>%
  mutate(tm = fct_relevel(tm, c("8hpi", "24hpi")))

# Create boxplots
C <- gns %>%
  map(~ {
    lvls <- dat %>%
      group_by(lec_type) %>%
      summarize(
        stats = list(boxplot.stats(!!sym(.x))),
        med   = map_dbl(stats, ~ .x$stats[3]),
        q3    = map_dbl(stats, ~ .x$stats[4]),
        q4    = map_dbl(stats, ~ .x$stats[5]),
        mx    = max(!!sym(.x))
      ) %>%
      arrange(med, q3, q4, mx) %>%
      pull(lec_type) %>%
      rev()
      
    dat %>%
      mutate(lec_type = fct_relevel(lec_type, lvls)) %>%
      ggplot(aes(lec_type, !!sym(.x), color = lec_type, fill = lec_type)) +
      geom_boxplot(
        size = 0.5, alpha = 0.5,
        outlier.size = 0.1, outlier.alpha = 1,
        position = position_dodge2(preserve = "single")
      ) +
      scale_color_manual(values = lec_type_cols) +
      scale_fill_manual(values = lec_type_cols) +
      scale_x_discrete(labels = type_lab_fn) +
      facet_wrap(~ tm, ncol = 1) +
      labs(y = str_c(.x, " expression")) +
      base_theme +
      theme(
        aspect.ratio    = 0.16,
        legend.position = "none",
        axis.title.x    = element_blank(),
        axis.text.x     = element_text(angle = 45, hjust = 1),
        axis.text.y     = element_text(size = txt_pt1)
      )
  }) %>%
  plot_grid(
    plotlist = .,
    align = "vh",
    axis  = "trbl",
    nrow  = 2
  )
```

```{r "fig S1", fig.width = 15, fig.height = 14}
top <- plot_grid(
  A, B,
  labels = c("A", "B"), label_size = ttl_pt2 * 1.5,
  align = "h", axis = "tb",
  nrow = 1
)

plot_grid(
  top, C,
  labels = c("", "C"), label_size = ttl_pt2 * 1.5,
  align = "v", axis = "rl",
  ncol = 1
)
```


```{r "cell type markers", fig.height = 21, fig.width = 20, eval = FALSE}
feats <- c(
  "Ptprc",           # hematopoietic cells
  "Pdpn", "Pecam1",  # LECs (+/+), BECs (-/+), FRCs (+/-)
  "Pdgfra",          # Fibroblasts
  "Acta2", "Itga7",  # PvCs
  "Cd14", "Fcgr3",   # macrophages/monocytes
  "Fcgr1",           # macrophages
  "Nkg7",            # NK cells
  "Cd3e",            # T cells
  "Cd19"             # B cells
)

feats %>%
  map(~ {
    so_8 %>%
      plot_example_gene(
        .x, "lec_type",
        box_colors  = lec_type_cols,
        umap_colors = c("white", "#D7301F")
      )
  }) %>%
  plot_grid(plotlist = ., ncol = 4)
```

```{r "lec type markers", fig.height = 28, fig.width = 20, eval = FALSE}
feats <- c(
  "Marco", "Lyve1", "Tspan7",        # Marco LECs
  "Madcam1", "Vcam1", "Clu", "Ltb",  # fLECs
  "Ptx3", "Stab2",                   # Ptx3 LECs
  "Cldn11", "Esam",                  # Valve LECs
  "Ackr4", "Foxc2",                  # cLECs
  "Ccl21a"                           # Collecting LECs
)

feats %>%
  map(~ {
    so_8 %>%
      plot_example_gene(
        .x, "lec_type",
        box_colors  = lec_type_cols,
        umap_colors = c("white", "#D7301F")
      )
  }) %>%
  plot_grid(plotlist = ., ncol = 4)
```

<br>

## Figure 5

CHIKV infection induces distinct cell type-specific gene expression programs

A. The average fold change in expression is shown below for the top 6 CHIKV-induced gene expression programs. Fold changes were scaled and centered for each gene (row). A darker tile indicates the gene is more strongly upregulated in the cell type (e.g. the Marco LEC cluster includes genes that are generally more highly upregulated in Marco LECs)

B. The top 6 gene ontology terms are shown for the gene expression programs described in A.

C. Select genes identified in A are shown below for mock- and CHIKV-infected samples. p-values comparing mock- and CHIKV-infected samples are shown above each set of boxplots.

Brief description of approach:

1. Upregulated genes for each cell type were identified by comparing mock- and CHIKV-infected samples. To be included, a gene must have an adjusted p-value <0.05 for all replicates and >1.25 fold increase in expression for all replicates.
2. Hierarchical clustering was used to divide upregulated genes into 10 groups based on their average fold change in expression for each cell type.
3. The top 7 clusters are shown below. The top 2 clusters (based on number of genes) appeared functionally similar and were merged into one cluster (interferon response).

```{r "GSEA MARKERS"}

# Cell types to test
typs <- so_24 %>%
  get_cell_types("lec_type", "orig.ident", n_cells = 3)

# Find marker genes for each cell type
marks_path <- here(params$so_dir, "all_markers.tsv")

if (!file.exists(marks_path)) {
  Idents(so_24) <- so_24 %>%
    FetchData("treatment")
  
  marks <- typs %>%
    map_dfr(~ {
      so_24 %>%
        subset(subset = lec_type == .x) %>%
        find_conserved_markers(
          ident_1  = "CHIKV",
          grp_var  = "rep",
          p_max    = 1.1,
          fc_range = c(-Inf, Inf)
        ) %>%
        mutate(cell_type = .x)
    }) %>%
    arrange(desc(avg_log2FC))
  
  marks %>%
    write_tsv(marks_path)
  
} else {
  marks <- read_tsv(marks_path)
}

# Exclude ribosomal protein and mitochondrial genes
marks <- marks %>%
  filter(!grepl("^Rp[sl]|^mt-", gene))

```

```{r "H CLUSTERING"}

# Marker stats columns
reps     <- unique(so_24$rep)
fc_clmns <- str_c(reps, "_avg_log2FC")
p_clmns  <- str_c(reps, "_p_val_adj")

# Format matrix
# Only include genes >1.25 fold change and <0.05 adj p-value
sig_marks <- marks %>%
  filter(
    if_all(all_of(fc_clmns), ~ .x > log2(1.25)),
    if_all(all_of(p_clmns),  ~ .x < 0.05)
  ) %>%
  rowwise() %>%
  mutate(max_p = max(!!!syms(p_clmns))) %>%
  ungroup() %>%
  dplyr::select(gene, cell_type, avg_log2FC, max_p)
  
mat <- sig_marks %>%
  dplyr::select(-max_p) %>%
  pivot_wider(
    names_from  = cell_type,
    values_from = avg_log2FC,
    values_fill = 0
  ) %>%
  column_to_rownames("gene") %>%
  t() %>%
  scale() %>%
  t()

# Hierarchical clustering
hcl <- mat %>%
  dist() %>%
  hclust(method = "ward.D")

clsts <- cutree(hcl, k = 10)

clsts <- tibble(
  gene    = names(clsts),
  cluster = as.character(unname(clsts))
)

# Summarize gene fold change for each cluster
# Calculate overall average fold change and p-value for each cell_type
clst_dat <- marks %>%
  dplyr::select(gene, cell_type, avg_log2FC, max_pval) %>%
  right_join(clsts, by = "gene") %>%
  group_by(cluster, cell_type) %>%
  mutate(
    avg_log2FC = mean(avg_log2FC),
    max_pval   = mean(max_pval),
  ) %>%
  ungroup()

# Only include cell type if meets FC and p threshold for the cluster
clst_dat <- clst_dat %>%
  group_by(cluster) %>%
  mutate(
    max_fc     = max(avg_log2FC),
    max_fc_lim = max(avg_log2FC[avg_log2FC < unique(max_fc)]) * 1.5
  ) %>%
  filter(
    cell_type %in% cell_type[avg_log2FC > log2(1.25) & max_pval < 0.05]
  )

# Further filter cell types for each cluster
# if one cell type is >1.5 fold higher than other cell types,
# remove other cell types
clst_dat <- clst_dat %>%
  filter(
    n_distinct(cell_type) == 1 |
    cell_type %in% cell_type[
      (avg_log2FC == max_fc & max_fc > max_fc_lim) |
      all(avg_log2FC <= max_fc_lim)
    ]
  ) %>%
  mutate(
    up_types = list(unique(cell_type)),
    n_types = n_distinct(cell_type)
  ) %>%
  ungroup() %>%
  mutate(shared = n_types >= 5)

# Merge clusters with >4 positive cell types
# Create data.frame with new cluster IDs
clst_dat <- clst_dat %>%
  mutate(cluster = ifelse(shared, max(cluster[shared]), cluster))

# Identify top clusters
up_types <- clst_dat %>%
  distinct(cluster, gene, up_types, n_types) %>%
  group_by(cluster, up_types, n_types) %>%
  summarize(n = n_distinct(gene), .groups = "drop")

top_clsts <- up_types %>%
  group_by(cluster, n_types, n) %>%
  summarize(id = map_chr(up_types, str_c, collapse = "/"), .groups = "drop") %>%
  mutate(id = ifelse(n_types >= 5, "interferon response", id)) %>%
  group_by(cluster, id) %>%
  summarize(n = sum(n), .groups = "drop") %>%
  arrange(desc(n)) %>%
  mutate(
    rnk = row_number(),
    id  = str_replace(id, "_", " "),
    id = str_c(id, "\n", n, " genes")
  ) %>%
  head(6)

clst_lvls <- set_names(top_clsts$id, top_clsts$cluster)
  
# Format scaled data for plotting
# Create new table with revised cluster IDs
clsts <- clst_dat %>%
  distinct(gene, cluster) %>%
  filter(cluster %in% top_clsts$cluster)

plt_dat <- marks %>%
  filter(
    gene %in% unique(sig_marks$gene),
    cell_type %in% colnames(mat)
  ) %>%
  
  dplyr::select(gene, cell_type, avg_log2FC) %>%
  pivot_wider(
    names_from  = cell_type,
    values_from = avg_log2FC,
    values_fill = 0
  ) %>%
  column_to_rownames("gene") %>%
  t() %>%
  scale() %>%
  t() %>%

  as_tibble(rownames = "gene") %>%
  left_join(clsts, by = "gene") %>%
  filter(!is.na(cluster)) %>%
  pivot_longer(-c(gene, cluster), names_to = "cell_type") %>%
  mutate(
    cell_type = as.factor(cell_type),
    cluster = fct_relevel(cluster, top_clsts$cluster)
  ) %>%
  arrange(cluster, cell_type)
  
# ORIGINAL VERSION
# plt_dat <- mat %>%
#   as_tibble(rownames = "gene") %>%
#   left_join(clsts, by = "gene") %>%
#   filter(!is.na(cluster)) %>%  # NAs are genes not found in any cluster
#   pivot_longer(-c(gene, cluster), names_to = "cell_type") %>%
#   mutate(
#     cluster = fct_relevel(cluster, top_clsts$cluster)
#   ) %>%
#   arrange(cluster, cell_type)

# Format data for drawing heatmap boxes
type_lvls <- levels(plt_dat$cell_type)
type_lvls <- set_names(seq_along(type_lvls), type_lvls)

bx_dat <- up_types %>%
  filter(n_types < 3) %>%
  unnest(up_types) %>%
  group_by(cluster, up_types) %>%
  mutate(
    x = type_lvls[up_types],
    x = list(c(x - 0.5, x + 0.5))
  ) %>%
  ungroup() %>%
  unnest(x) %>%
  filter(cluster %in% plt_dat$cluster) %>%
  mutate(cluster = fct_relevel(cluster, levels(plt_dat$cluster)))

# Create heatmaps
clst_clrs <- c(
  "3" = lec_type_cols[["unassigned-LEC"]],
  "8" = lec_type_cols[["BEC"]],
  "5" = lec_type_cols[["PvC"]],
  "2" = lec_type_cols[["Macrophages"]],
  "4" = lec_type_cols[["Marco_LEC"]],
  "9" = lec_type_cols[["FRC"]]
)

h_heat <- plt_dat %>%
  # ggplot(aes(cell_type, gene, fill = value)) +
  # scale_fill_gradient2(low = "grey85", mid = "white", high = "#D7301F") +
  # scale_fill_gradient2(low = "#56B4E9", mid = "white", high = "#D7301F") +
  # guides(fill = guide_colorbar(
  #   ticks = FALSE,
  #   title.position = "bottom",
  #   barwidth = unit(150, "pt"), barheight = unit(10, "pt")
  # )) +
  
  ggplot(aes(cell_type, gene, fill = cluster, alpha = value)) +
  scale_alpha_continuous(range = c(0.01, 1)) +
  scale_fill_manual(values = clst_clrs) +
  guides(
    fill = "none",
    alpha = guide_legend(
      label.position = "bottom",
      keyheight = unit(7, "pt"),
      title = "z-score",
      title.position = "top"
    )
  ) +
  geom_vline(data = bx_dat, aes(xintercept = x)) +
  geom_tile() +
  facet_wrap(
    ~ cluster,
    scales         = "free_y",
    strip.position = "left",
    labeller       = as_labeller(clst_lvls),
    ncol = 1
  ) +
  scale_x_discrete(position = "top") +
  djvdj_theme() +
  theme(
    axis.text.y  = element_blank(),
    axis.ticks.y = element_blank(),
    axis.text.x  = element_text(angle = 90, hjust = 0, vjust = 0.5),
    axis.title   = element_blank(),
    legend.position = "bottom",
    legend.title    = element_text(size = txt_pt1, hjust = 0.5),
    strip.clip = "off"
  )

```

```{r "H CLUSTERING GO"}

# Format genes for GO analysis
go_gns <- clsts %>%
  split(as.character(.$cluster)) %>%
  map(pull, gene)

# Identify GO terms
go_res <- go_gns %>%
  gost(
    organism = "mmusculus",
    ordered_query = FALSE,
    evcodes = TRUE
  ) %>%
  .$result %>%
  filter(
    source %in% c("GO:BP", "GO:MF", "KEGG"),
    p_value < 0.05,
    term_size < 500
  ) %>%
  arrange(p_value)

# Create GO bargraphs
lab_fn <- function(x) {
  x %>%
    str_remove("^[0-9]+-") %>%
    str_trunc(80) %>%
    str_wrap(45)
}

go_bars <- go_res %>%
  group_by(query) %>%
  dplyr::slice(1:6) %>%
  ungroup() %>%
  dplyr::select(-parents) %>%
  mutate(
    term_name = str_c(query, "-", term_name),
    term_name = fct_reorder(term_name, p_value, .desc = TRUE),
    query     = fct_relevel(query, names(clst_lvls))
  ) %>%
  ggplot(aes(-log10(p_value), term_name, fill = query)) +
  geom_col() +
  labs(x = "-log10(p-value)") +
  facet_wrap(~ query, scales = "free_y", strip.position = "right", ncol = 1) +
  scale_fill_manual(values = clst_clrs) +
  scale_y_discrete(labels = lab_fn) +
  djvdj_theme() +
  theme(
    axis.title.y = element_blank(),
    axis.title.x = element_text(size = txt_pt1),
    legend.position = "none",
    strip.text = element_blank()
  )

go_res %>%
  write_tsv(file.path(params$so_dir, "gene_clusters_go.tsv"))

```

```{r "H CLUSTERING TOP GENES"}

# gns <- c("Cd74", "H2-Ab1", "Lgmn", "Angpt2", "Pik3r1", "Gja1", "Ywhag")    # BEC
# gns <- c("Vcam1", "Akt3", "Cdh5", "Ccl5", "F11r", "Olr1", "Dst", "Nlgn2")  # Marco
# gns <- c("Cxcl9", "Cxcl10", "Ccl2", "Ccl7", "Ccl4", "Il1b", "Il6", "Tnf")  # All
# gns <- c("Nod1", "Tlr3")                                                   # FRC/PvC
# gns <- c("Oas2", "Oas3", "Fgl2", "Irf1", "Cxcl10")                         # Mac

gns <- c("Irf7", "Ccl2", "Cxcl9", "Oas3", "Vcam1", "Tlr3")

# Format plotting data
dat <- so_24 %>%
  FetchData(c("treatment", "orig.ident", "lec_type", gns)) %>%
  rownames_to_column("cell_id") %>%
  pivot_longer(all_of(gns), names_to = "gene") %>%
  filter(lec_type %in% colnames(mat)) %>%
  mutate(
    treatment = fct_relevel(treatment, treats),
    gene = fct_relevel(gene, gns)
  )
  
# Format data for p-value labels
p_dat <- sig_marks %>%
  mutate(p_lab = format_pvalue(max_p)) %>%
  dplyr::rename(lec_type = cell_type) %>%
  filter(gene %in% gns) %>%
  mutate(gene = fct_relevel(gene, gns))

# Create boxplots
bx_fn <- function(x) str_c(x, "\nexpression")

gn_bxs <- dat %>%
  ggplot(aes(lec_type, value, alpha = treatment, fill = lec_type, color = lec_type)) +
  
  geom_boxplot(
    outlier.size = 0.001,
    position = position_dodge2(preserve = "single"),
    key_glyph = draw_key_point
  ) +

  geom_richtext(
    aes(lec_type, Inf, label = p_lab),
    data = p_dat,
    alpha = 1,
    size  = 5 / .pt,
    color = "black",
    fill  = NA,
    label.color = NA,
    vjust = "inward"
  ) +
    
  facet_wrap(
    ~ gene,
    scales = "free_y",
    ncol = 1,
    strip.position = "left",
    labeller = as_labeller(bx_fn)
  ) +
  guides(
    fill = "none",
    color = "none",
    alpha = guide_legend(override.aes = list(shape = 15, alpha = c(0.5, 1), size = 4))
  ) +
  scale_y_continuous(expand = expansion(c(0.05, 0.2))) +
  scale_alpha_manual(values = c(0.1, 0.6)) +
  scale_fill_manual(values = lec_type_cols) +
  scale_color_manual(values = lec_type_cols) +
  djvdj_theme() +
  theme(
    axis.title      = element_blank(),
    axis.text.x     = element_text(angle = 45, hjust = 1),
    legend.title    = element_blank(),
    legend.position = "top",
    strip.placement = "outside"
    # plot.margin = margin(15, 15, 15, 30),
  )

```

```{r "H FINAL", fig.height = 10, fig.width = 10}

plot_grid(
  h_heat, go_bars, gn_bxs,
  align = "h",
  axis  = "tb",
  nrow  = 1,
  rel_widths = c(0.75, 1, 1),
  labels = c("A", "B", "C"),
  label_size = ttl_pt2 * 1.5
)

```



```{r "GSEA RESULTS", eval = FALSE}

# Gene sets to use
m_db <- msigdbr(species = "Mus musculus")

to_query <- c("H", "C2", "C3", "C5")

gene_set <- m_db %>%
  filter(gs_cat %in% to_query) %>%
  dplyr::select(gs_name, gene_symbol)

# Turn off BiocParallel multi-threading
# Memory usage exceeds 32G when using 6 threads
options(MulticoreParam = BiocParallel::MulticoreParam(workers = 1))

# Run GSEA
gsea_path <- here(params$so_dir, "all_gsea.tsv")

if (!file.exists(gsea_path)) {
  gsea_res <- set_names(typs) %>%
    map_dfr(~ {
      gene_lst <- filter(marks, cell_type == .x)
      gene_lst <- set_names(gene_lst$avg_log2FC, gene_lst$gene)
      
      set.seed(42)
      
      gene_lst %>%
        GSEA(
          TERM2GENE    = gene_set,
          pvalueCutoff = 0.05,
          eps          = 1e-50
        ) %>%
        as_tibble() %>%
        mutate(cell_type = .x)
    })
  
  gsea_res %>%
    write_tsv(here(params$so_dir, "gsea.tsv"))
  
} else {
  gsea_res <- read_tsv(gsea_path)
}

# Focus on enriched pathway terms
gsea_res <- gsea_res %>%
  filter(
    enrichmentScore > 0,
    grepl("^(HALLMARK|WP|KEGG)_", ID)
  )

```

```{r "GSEA PLOTS", eval = FALSE}

k_dat <- gsea_res %>%
  group_by(cell_type) %>%
  dplyr::slice(1:30) %>%
  dplyr::select(ID, enrichmentScore, cell_type) %>%
  pivot_wider(
    names_from  = cell_type,
    values_from = enrichmentScore,
    values_fill = 0
  ) %>%
  column_to_rownames("ID")

k_res <- k_dat %>%
  kmeans(6)

# Bar graphs
plt_dat <- gsea_res %>%
  mutate(ID = fct_reorder(ID, -log10(p.adjust), mean))
  
plt_dat %>%
  # group_by(cell_type) %>%
  # dplyr::slice(1:10) %>%
  # ungroup() %>%
  filter(grepl("^(HALLMARK)_", ID)) %>%
  ggplot(aes(-log10(p.adjust), ID, fill = cell_type)) +
  geom_col() +
  facet_wrap(~ cell_type, ncol = 1, scales = "free_y") +
  scale_fill_manual(values = lec_type_cols) +
  djvdj_theme()

# Heatmaps
plt_dat %>%
  # group_by(cell_type) %>%
  # dplyr::slice(1:20) %>%
  # ungroup() %>%
  filter(grepl("^(HALLMARK|KEGG)_", ID)) %>%
  ggplot(aes(cell_type, ID, fill = -log10(p.adjust))) +
  geom_tile() +
  scale_fill_gradientn(colours = c("lightblue", "red")) +
  djvdj_theme() +
  scale_x_discrete(position = "top") +
  theme(
    axis.title.x = element_blank(),
    axis.text.x  = element_text(angle = 45, hjust = 0)
  )

```

```{r "GSEA MARKER HEATMAP", eval = FALSE}

# Fold changes for every gene
fc_dat <- marks %>%
  left_join(m_db, by = c("gene" = "gene_symbol")) %>%
  dplyr::select(gs_cat, gs_name, gene, avg_log2FC, cell_type) %>%
  filter(grepl("^(HALLMARK|KEGG)_", gs_name))

# Format GSEA data
terms <- gsea_res %>%
  filter(grepl("^(HALLMARK|KEGG)_", ID)) %>%
  group_by(cell_type) %>%
  dplyr::slice(1:10) %>%
  ungroup() %>%
  separate_rows(core_enrichment, sep = "/")

# Identify top genes
gns <- terms %>%
  group_by(cell_type, core_enrichment) %>%
  summarize(n = n(), .groups = "drop") %>%
  group_by(cell_type) %>%
  slice_max(n, n = 5)

# Identify top terms
# Overall top terms
# top_terms <- terms %>%
#   distinct(ID, cell_type, p.adjust) %>%
#   group_by(ID) %>%
#   summarize(
#     p.adjust = mean(p.adjust),
#     n = n_distinct(cell_type)
#   ) %>%
#   arrange(desc(n)) %>%
#   pull(ID)
top_terms <- terms %>%
  distinct(cell_type, ID)

# Format plot data
plt_dat <- fc_dat %>%
  filter(
    # gs_name %in% top_terms,
    gene %in% gns$core_enrichment
  ) %>%
  semi_join(top_terms, by = c(gs_name = "ID", "cell_type"))

# Add GSEA p-values
plt_dat <- terms %>%
  distinct(ID, p.adjust, cell_type) %>%
  right_join(plt_dat, by = c(ID = "gs_name", "cell_type"))

plt_dat %>%
  mutate(
    # ID   = fct_relevel(ID, rev(top_terms)),
    ID   = fct_reorder(ID, p.adjust, .desc = TRUE),
    gene = fct_reorder(gene, avg_log2FC, .desc = TRUE)
  ) %>%

# plt_dat %>%
#   filter(core_enrichment %in% unique(gns$core_enrichment)) %>%
#   group_by(core_enrichment) %>%
#   mutate(n = n_distinct(ID)) %>%
#   ungroup() %>%
#   mutate(core_enrichment = fct_reorder(core_enrichment, n, .desc = TRUE)) %>%
  
  ggplot(aes(gene, ID, fill = avg_log2FC)) +
  geom_tile() +
  facet_wrap(~ cell_type, ncol = 2, scales = "free_y") +
  scale_fill_gradient2(low = "blue", mid = "grey90", high = "red") +
  djvdj_theme() +
  theme(
    axis.title = element_blank(),
    axis.text.x = element_text(angle = 45, hjust = 1)
  )

```

```{r "K CLUSTERING DEGs", eval = FALSE}

reps     <- unique(so_24$rep)
fc_clmns <- str_c(reps, "_avg_log2FC")
p_clmns  <- str_c(reps, "_p_val_adj")

# Cluster genes based on FC for each cell type
# degs <- marks %>%
#   filter(
#     if_all(all_of(fc_clmns), ~ .x > 0),
#     if_all(all_of(p_clmns),  ~ .x < 0.05)
#   ) %>%
#   pull(gene) %>%
#   unique()

k_dat <- marks %>%
  filter(
    if_all(all_of(fc_clmns), ~ .x > 0),
    if_all(all_of(p_clmns),  ~ .x < 0.05)
  ) %>%
  # filter(gene %in% degs) %>%
  dplyr::select(gene, avg_log2FC, cell_type) %>%
  pivot_wider(
    names_from  = cell_type,
    values_from = avg_log2FC,
    values_fill = 0
  ) %>%
  column_to_rownames("gene") %>%
  t() %>%
  scale() %>%
  t()

k_res <- k_dat %>%
  kmeans(20)

k_res <- tibble(
  gene    = names(k_res$cluster),
  cluster = unname(k_res$cluster)
)

# Format plot data
clsts <- unique(k_res$cluster)

clsts %>%
  map(~ {
    # marks %>%
    k_dat %>%
      as.data.frame() %>%
      rownames_to_column("gene") %>%
      pivot_longer(-gene, names_to = "cell_type", values_to = "avg_log2FC") %>%
      inner_join(k_res, by = "gene") %>%
      filter(cluster == .x) %>%
      
      group_by(gene) %>%
      mutate(avg_log2FC = as.numeric(scale(avg_log2FC))) %>%
      ungroup() %>%
      
      ggplot(aes(cell_type, gene, fill = avg_log2FC)) +
      geom_tile() +
      scale_fill_gradient2(low = "blue", mid = "white", high = "red") +
      facet_wrap(~ cluster, scales = "free") +
      djvdj_theme() +
      theme(
        axis.text.y  = element_blank(),
        axis.ticks.y = element_blank(),
        axis.text.x  = element_text(angle = 45, hjust = 1),
        axis.title   = element_blank()
      )
  }) %>%
  plot_grid(plotlist = .)

```

```{r "Cell type CellChat", eval = FALSE}

# Create CellChat objects
prfx     <- "all_"
typ_clmn <- "cell_type"
clmns    <- c("orig.ident", "treatment", "rep", typ_clmn)

typ_cc <- so_24 %>%
  run_cellchat(
    group_col    = "treatment",
    cell_col     = typ_clmn,
    include_cols = clmns,
    pos_group    = "CHIKV",
    prefix       = prfx,
    object_dir   = params$so_dir
  )

# Table of differential pathways
typ_net <- typ_cc$cellchat %>%
  netMappingDEG(
    features.name = "CHIKV",
    thresh = 1
  )

typ_net_up <- typ_cc$cellchat %>%
  subsetCommunication(
    net            = typ_net,
    datasets       = "CHIKV",
    ligand.logFC   = 0.2,
    receptor.logFC = NULL
  ) %>%
  filter(pval < 0.05)
  # filter(ligand.pvalues < 0.05)

```

```{r "LEC CellChat", eval = FALSE}

# Cell types to includes
# only include ones with >30 cells for mock and CHIKV
prfx     <- "lec_"
typ_clmn <- str_c(prfx, "type")
clmns    <- c("orig.ident", "treatment", "rep", "cell_type", typ_clmn)

typs <- so_24@meta.data %>%
  group_by(treatment, !!sym(typ_clmn)) %>%
  summarize(n = n(), .groups = "drop") %>%
  filter(n > 30) %>%
  group_by(!!sym(typ_clmn)) %>%
  filter(all(treats %in% treatment)) %>%
  ungroup() %>%
  pull(lec_type) %>%
  unique()

# Create CellChat objects
lec_cc <- so_24 %>%
  subset(!!sym(typ_clmn) %in% typs) %>%
  run_cellchat(
    group_col    = "treatment",
    cell_col     = typ_clmn,
    include_cols = clmns,
    pos_group    = "CHIKV",
    prefix       = prfx,
    object_dir   = params$so_dir
  )

# Table of differential pathways
lec_net <- lec_cc$cellchat %>%
  netMappingDEG(
    features.name = "CHIKV",
    thresh = 1
  )

lec_net_up <- lec_cc$cellchat %>%
  subsetCommunication(
    net            = lec_net,
    datasets       = "CHIKV",
    ligand.logFC   = 0.2,
    receptor.logFC = NULL
  ) %>%
  filter(
    ligand.pvalues < 0.05,
    ligand.logFC > 0.5,
    receptor.logFC > 0
  )

```

```{r "CHIKV TOTAL INTERACTIONS HEATMAP", fig.width = 6, fig.height = 5, eval = FALSE}

# All cell types
file.path(params$so_dir, "all_chikv_heat.png") %>%
  png(res = 300, width = 6, height = 5, units = "in")

typ_cc$cc_objs$CHIKV %>%
  netVisual_heatmap(color.heatmap = "Reds", measure = "weight")

dev.off()

# LEC types
file.path(params$so_dir, "lec_chikv_heat.png") %>%
  png(res = 300, width = 6, height = 5, units = "in")

lec_cc$cc_objs$CHIKV %>%
  netVisual_heatmap(color.heatmap = "Reds", measure = "weight")

dev.off()

```

```{r "DIFF INTERACTIONS HEATMAP", fig.width = 6, fig.height = 5, eval = FALSE}

# All cell types
file.path(params$so_dir, "all_diff_heat.png") %>%
  png(res = 300, width = 6, height = 5, units = "in")

typ_cc$cellchat %>%
  netVisual_heatmap(measure = "weight")

dev.off()

# LEC types
file.path(params$so_dir, "lec_diff_heat.png") %>%
  png(res = 300, width = 6, height = 5, units = "in")

lec_cc$cellchat %>%
  netVisual_heatmap(measure = "weight")

dev.off()

```

```{r "GO", eval = FALSE}

# Find marker genes
Idents(so_24) <- so_24$treatment

marks <- typs %>%
  map_dfr(~ {
    so_24 %>%
      subset(lec_type == .x) %>%
      FindConservedMarkers(
        ident.1 = "CHIKV",
        ident.2 = "mock",
        grouping.var = "rep",
        only.pos = TRUE
      ) %>%
      as_tibble(rownames = "gene") %>%
      mutate(cell_type = .x)
  })

# Filter markers
reps <- unique(so_24$rep)

fc_clmns   <- str_c(reps, "_avg_log2FC")
p_clmns    <- str_c(reps, "_p_val_adj")
pct1_clmns <- str_c(reps, "_pct.1")
pct2_clmns <- str_c(reps, "_pct.2")

up <- marks %>%
  filter(
    if_all(all_of(fc_clmns),   ~ .x > log2(1.25)),
    if_all(all_of(p_clmns),    ~ .x < 0.05),
    if_all(all_of(pct1_clmns), ~ .x > 0.3),
    if_all(all_of(pct2_clmns), ~ .x < 0.5)
  )

# Identify GO terms
go <- up %>%
  split(.$cell_type) %>%
  imap_dfr(~ {
    bg <- so_24 %>%
      subset(lec_type == .y) %>%
      subset(treatment == treats[1])
    
    bg <- bg@assays$RNA@data %>%
      rowMeans() %>%
      sort(decreasing = TRUE) %>%
      head(5000) %>%
      names()
    
    res <- gost(
      query        = .x$gene,
      organism     = "mmusculus",
      domain_scope = "custom",
      custom_bg    = bg,
      evcodes      = TRUE,
      significant  = TRUE
    ) %>%
      .$result
    
    if (is_empty(res)) return(NULL)
    
    res %>%
      as_tibble() %>%
      mutate(query = .y) %>%
      dplyr::select(-evidence_codes, -parents)
  })

# Plot top GO terms
top_go <- go %>%
  filter(
    term_size > 10, term_size < 400,
    intersection_size > 5,
    source %in% c("GO:BP", "HP", "KEGG", "WP")
  ) %>%
  rowwise() %>%
  mutate(
    intersection = list(str_split(intersection, ",")[[1]]),
    frac_term = intersection_size / term_size
  ) %>%
  ungroup() %>%
  arrange(query, p_value)

top_go %>%
  group_by(query) %>%
  slice(1:10) %>%
  
  ggplot(aes(-log10(p_value), term_name, size = frac_term, color = query)) +
  geom_point() +
  facet_wrap(~ query) +
  scale_color_manual(values = lec_type_cols) +
  djvdj_theme()
  
```

```{r "GSEA", eval = FALSE}


```
